// svix/dist/api/application.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Application = void 0;
const applicationIn_1 = require("../models/applicationIn");
const applicationOut_1 = require("../models/applicationOut");
const applicationPatch_1 = require("../models/applicationPatch");
const listResponseApplicationOut_1 = require("../models/listResponseApplicationOut");
const request_1 = require("../request");
class Application {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    list(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app");
            request.setQueryParams({
                exclude_apps_with_no_endpoints: options === null || options === void 0 ? void 0 : options.excludeAppsWithNoEndpoints,
                exclude_apps_with_disabled_endpoints: options === null || options === void 0 ? void 0 : options.excludeAppsWithDisabledEndpoints,
                exclude_apps_with_svix_play_endpoints: options === null || options === void 0 ? void 0 : options.excludeAppsWithSvixPlayEndpoints,
                limit: options === null || options === void 0 ? void 0 : options.limit,
                iterator: options === null || options === void 0 ? void 0 : options.iterator,
                order: options === null || options === void 0 ? void 0 : options.order,
            });
            return yield request.send(this.requestCtx, listResponseApplicationOut_1.ListResponseApplicationOutSerializer._fromJsonObject);
        });
    }
    create(applicationIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app");
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(applicationIn_1.ApplicationInSerializer._toJsonObject(applicationIn));
            return yield request.send(this.requestCtx, applicationOut_1.ApplicationOutSerializer._fromJsonObject);
        });
    }
    getOrCreate(applicationIn, options) {
        const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app");
        request.setQueryParam("get_if_exists", true);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        request.setBody(applicationIn_1.ApplicationInSerializer._toJsonObject(applicationIn));
        return request.send(this.requestCtx, applicationOut_1.ApplicationOutSerializer._fromJsonObject);
    }
    get(appId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}");
            request.setPathParam("app_id", appId);
            return yield request.send(this.requestCtx, applicationOut_1.ApplicationOutSerializer._fromJsonObject);
        });
    }
    update(appId, applicationIn) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/app/{app_id}");
            request.setPathParam("app_id", appId);
            request.setBody(applicationIn_1.ApplicationInSerializer._toJsonObject(applicationIn));
            return yield request.send(this.requestCtx, applicationOut_1.ApplicationOutSerializer._fromJsonObject);
        });
    }
    delete(appId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}");
            request.setPathParam("app_id", appId);
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    patch(appId, applicationPatch) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}");
            request.setPathParam("app_id", appId);
            request.setBody(applicationPatch_1.ApplicationPatchSerializer._toJsonObject(applicationPatch));
            return yield request.send(this.requestCtx, applicationOut_1.ApplicationOutSerializer._fromJsonObject);
        });
    }
}
exports.Application = Application;
