// svix/dist/api/authentication.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Authentication = void 0;
const apiTokenOut_1 = require("../models/apiTokenOut");
const appPortalAccessIn_1 = require("../models/appPortalAccessIn");
const appPortalAccessOut_1 = require("../models/appPortalAccessOut");
const applicationTokenExpireIn_1 = require("../models/applicationTokenExpireIn");
const rotatePollerTokenIn_1 = require("../models/rotatePollerTokenIn");
const streamPortalAccessIn_1 = require("../models/streamPortalAccessIn");
const streamTokenExpireIn_1 = require("../models/streamTokenExpireIn");
const dashboardAccessOut_1 = require("../models/dashboardAccessOut");
const request_1 = require("../request");
class Authentication {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    appPortalAccess(appId, appPortalAccessIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/app-portal-access/{app_id}");
            request.setPathParam("app_id", appId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(appPortalAccessIn_1.AppPortalAccessInSerializer._toJsonObject(appPortalAccessIn));
            return yield request.send(this.requestCtx, appPortalAccessOut_1.AppPortalAccessOutSerializer._fromJsonObject);
        });
    }
    expireAll(appId, applicationTokenExpireIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/app/{app_id}/expire-all");
            request.setPathParam("app_id", appId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(applicationTokenExpireIn_1.ApplicationTokenExpireInSerializer._toJsonObject(applicationTokenExpireIn));
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    dashboardAccess(appId, options) {
        const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/dashboard-access/{app_id}");
        request.setPathParam("app_id", appId);
        request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
        return request.send(this.requestCtx, dashboardAccessOut_1.DashboardAccessOutSerializer._fromJsonObject);
    }
    logout(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/logout");
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    streamLogout(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream-logout");
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    streamPortalAccess(streamId, streamPortalAccessIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream-portal-access/{stream_id}");
            request.setPathParam("stream_id", streamId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(streamPortalAccessIn_1.StreamPortalAccessInSerializer._toJsonObject(streamPortalAccessIn));
            return yield request.send(this.requestCtx, appPortalAccessOut_1.AppPortalAccessOutSerializer._fromJsonObject);
        });
    }
    streamExpireAll(streamId, streamTokenExpireIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream/{stream_id}/expire-all");
            request.setPathParam("stream_id", streamId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(streamTokenExpireIn_1.StreamTokenExpireInSerializer._toJsonObject(streamTokenExpireIn));
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    getStreamPollerToken(streamId, sinkId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/auth/stream/{stream_id}/sink/{sink_id}/poller/token");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            return yield request.send(this.requestCtx, apiTokenOut_1.ApiTokenOutSerializer._fromJsonObject);
        });
    }
    rotateStreamPollerToken(streamId, sinkId, rotatePollerTokenIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/auth/stream/{stream_id}/sink/{sink_id}/poller/token/rotate");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(rotatePollerTokenIn_1.RotatePollerTokenInSerializer._toJsonObject(rotatePollerTokenIn));
            return yield request.send(this.requestCtx, apiTokenOut_1.ApiTokenOutSerializer._fromJsonObject);
        });
    }
}
exports.Authentication = Authentication;
