// svix/dist/api/backgroundTask.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackgroundTask = void 0;
const backgroundTaskOut_1 = require("../models/backgroundTaskOut");
const listResponseBackgroundTaskOut_1 = require("../models/listResponseBackgroundTaskOut");
const request_1 = require("../request");
class BackgroundTask {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    list(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/background-task");
            request.setQueryParams({
                status: options === null || options === void 0 ? void 0 : options.status,
                task: options === null || options === void 0 ? void 0 : options.task,
                limit: options === null || options === void 0 ? void 0 : options.limit,
                iterator: options === null || options === void 0 ? void 0 : options.iterator,
                order: options === null || options === void 0 ? void 0 : options.order,
            });
            return yield request.send(this.requestCtx, listResponseBackgroundTaskOut_1.ListResponseBackgroundTaskOutSerializer._fromJsonObject);
        });
    }
    listByEndpoint(options) {
        return this.list(options);
    }
    get(taskId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/background-task/{task_id}");
            request.setPathParam("task_id", taskId);
            return yield request.send(this.requestCtx, backgroundTaskOut_1.BackgroundTaskOutSerializer._fromJsonObject);
        });
    }
}
exports.BackgroundTask = BackgroundTask;
