// svix/dist/api/connector.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Connector = void 0;
const connectorIn_1 = require("../models/connectorIn");
const connectorOut_1 = require("../models/connectorOut");
const connectorPatch_1 = require("../models/connectorPatch");
const connectorUpdate_1 = require("../models/connectorUpdate");
const listResponseConnectorOut_1 = require("../models/listResponseConnectorOut");
const request_1 = require("../request");
class Connector {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    list(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/connector");
            request.setQueryParams({
                limit: options === null || options === void 0 ? void 0 : options.limit,
                iterator: options === null || options === void 0 ? void 0 : options.iterator,
                order: options === null || options === void 0 ? void 0 : options.order,
                product_type: options === null || options === void 0 ? void 0 : options.productType,
            });
            return yield request.send(this.requestCtx, listResponseConnectorOut_1.ListResponseConnectorOutSerializer._fromJsonObject);
        });
    }
    create(connectorIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/connector");
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(connectorIn_1.ConnectorInSerializer._toJsonObject(connectorIn));
            return yield request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
        });
    }
    get(connectorId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/connector/{connector_id}");
            request.setPathParam("connector_id", connectorId);
            return yield request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
        });
    }
    update(connectorId, connectorUpdate) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/connector/{connector_id}");
            request.setPathParam("connector_id", connectorId);
            request.setBody(connectorUpdate_1.ConnectorUpdateSerializer._toJsonObject(connectorUpdate));
            return yield request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
        });
    }
    delete(connectorId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/connector/{connector_id}");
            request.setPathParam("connector_id", connectorId);
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    patch(connectorId, connectorPatch) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/connector/{connector_id}");
            request.setPathParam("connector_id", connectorId);
            request.setBody(connectorPatch_1.ConnectorPatchSerializer._toJsonObject(connectorPatch));
            return yield request.send(this.requestCtx, connectorOut_1.ConnectorOutSerializer._fromJsonObject);
        });
    }
}
exports.Connector = Connector;
