// svix/dist/api/endpoint.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Endpoint = void 0;
const bulkReplayIn_1 = require("../models/bulkReplayIn");
const endpointHeadersIn_1 = require("../models/endpointHeadersIn");
const endpointHeadersOut_1 = require("../models/endpointHeadersOut");
const endpointHeadersPatchIn_1 = require("../models/endpointHeadersPatchIn");
const endpointIn_1 = require("../models/endpointIn");
const endpointOut_1 = require("../models/endpointOut");
const endpointPatch_1 = require("../models/endpointPatch");
const endpointSecretOut_1 = require("../models/endpointSecretOut");
const endpointSecretRotateIn_1 = require("../models/endpointSecretRotateIn");
const endpointStats_1 = require("../models/endpointStats");
const endpointTransformationIn_1 = require("../models/endpointTransformationIn");
const endpointTransformationOut_1 = require("../models/endpointTransformationOut");
const endpointTransformationPatch_1 = require("../models/endpointTransformationPatch");
const endpointUpdate_1 = require("../models/endpointUpdate");
const eventExampleIn_1 = require("../models/eventExampleIn");
const listResponseEndpointOut_1 = require("../models/listResponseEndpointOut");
const messageOut_1 = require("../models/messageOut");
const recoverIn_1 = require("../models/recoverIn");
const recoverOut_1 = require("../models/recoverOut");
const replayIn_1 = require("../models/replayIn");
const replayOut_1 = require("../models/replayOut");
const request_1 = require("../request");
class Endpoint {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    list(appId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint");
            request.setPathParam("app_id", appId);
            request.setQueryParams({
                limit: options === null || options === void 0 ? void 0 : options.limit,
                iterator: options === null || options === void 0 ? void 0 : options.iterator,
                order: options === null || options === void 0 ? void 0 : options.order,
            });
            return yield request.send(this.requestCtx, listResponseEndpointOut_1.ListResponseEndpointOutSerializer._fromJsonObject);
        });
    }
    create(appId, endpointIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint");
            request.setPathParam("app_id", appId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(endpointIn_1.EndpointInSerializer._toJsonObject(endpointIn));
            return yield request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
        });
    }
    get(appId, endpointId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            return yield request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
        });
    }
    update(appId, endpointId, endpointUpdate) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setBody(endpointUpdate_1.EndpointUpdateSerializer._toJsonObject(endpointUpdate));
            return yield request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
        });
    }
    delete(appId, endpointId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    patch(appId, endpointId, endpointPatch) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setBody(endpointPatch_1.EndpointPatchSerializer._toJsonObject(endpointPatch));
            return yield request.send(this.requestCtx, endpointOut_1.EndpointOutSerializer._fromJsonObject);
        });
    }
    bulkReplay(appId, endpointId, bulkReplayIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/bulk-replay");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(bulkReplayIn_1.BulkReplayInSerializer._toJsonObject(bulkReplayIn));
            return yield request.send(this.requestCtx, replayOut_1.ReplayOutSerializer._fromJsonObject);
        });
    }
    getHeaders(appId, endpointId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            return yield request.send(this.requestCtx, endpointHeadersOut_1.EndpointHeadersOutSerializer._fromJsonObject);
        });
    }
    updateHeaders(appId, endpointId, endpointHeadersIn) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setBody(endpointHeadersIn_1.EndpointHeadersInSerializer._toJsonObject(endpointHeadersIn));
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    headersUpdate(appId, endpointId, endpointHeadersIn) {
        return this.updateHeaders(appId, endpointId, endpointHeadersIn);
    }
    patchHeaders(appId, endpointId, endpointHeadersPatchIn) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/headers");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setBody(endpointHeadersPatchIn_1.EndpointHeadersPatchInSerializer._toJsonObject(endpointHeadersPatchIn));
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    headersPatch(appId, endpointId, endpointHeadersPatchIn) {
        return this.patchHeaders(appId, endpointId, endpointHeadersPatchIn);
    }
    recover(appId, endpointId, recoverIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/recover");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(recoverIn_1.RecoverInSerializer._toJsonObject(recoverIn));
            return yield request.send(this.requestCtx, recoverOut_1.RecoverOutSerializer._fromJsonObject);
        });
    }
    replayMissing(appId, endpointId, replayIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/replay-missing");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(replayIn_1.ReplayInSerializer._toJsonObject(replayIn));
            return yield request.send(this.requestCtx, replayOut_1.ReplayOutSerializer._fromJsonObject);
        });
    }
    getSecret(appId, endpointId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/secret");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            return yield request.send(this.requestCtx, endpointSecretOut_1.EndpointSecretOutSerializer._fromJsonObject);
        });
    }
    rotateSecret(appId, endpointId, endpointSecretRotateIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/secret/rotate");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(endpointSecretRotateIn_1.EndpointSecretRotateInSerializer._toJsonObject(endpointSecretRotateIn));
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    sendExample(appId, endpointId, eventExampleIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/send-example");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(eventExampleIn_1.EventExampleInSerializer._toJsonObject(eventExampleIn));
            return yield request.send(this.requestCtx, messageOut_1.MessageOutSerializer._fromJsonObject);
        });
    }
    getStats(appId, endpointId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/stats");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setQueryParams({
                since: options === null || options === void 0 ? void 0 : options.since,
                until: options === null || options === void 0 ? void 0 : options.until,
            });
            return yield request.send(this.requestCtx, endpointStats_1.EndpointStatsSerializer._fromJsonObject);
        });
    }
    transformationGet(appId, endpointId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            return yield request.send(this.requestCtx, endpointTransformationOut_1.EndpointTransformationOutSerializer._fromJsonObject);
        });
    }
    patchTransformation(appId, endpointId, endpointTransformationPatch) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setBody(endpointTransformationPatch_1.EndpointTransformationPatchSerializer._toJsonObject(endpointTransformationPatch));
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    transformationPartialUpdate(appId, endpointId, endpointTransformationIn) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/app/{app_id}/endpoint/{endpoint_id}/transformation");
            request.setPathParam("app_id", appId);
            request.setPathParam("endpoint_id", endpointId);
            request.setBody(endpointTransformationIn_1.EndpointTransformationInSerializer._toJsonObject(endpointTransformationIn));
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
}
exports.Endpoint = Endpoint;
