// svix/dist/api/ingestEndpoint.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IngestEndpoint = void 0;
const ingestEndpointHeadersIn_1 = require("../models/ingestEndpointHeadersIn");
const ingestEndpointHeadersOut_1 = require("../models/ingestEndpointHeadersOut");
const ingestEndpointIn_1 = require("../models/ingestEndpointIn");
const ingestEndpointOut_1 = require("../models/ingestEndpointOut");
const ingestEndpointSecretIn_1 = require("../models/ingestEndpointSecretIn");
const ingestEndpointSecretOut_1 = require("../models/ingestEndpointSecretOut");
const ingestEndpointTransformationOut_1 = require("../models/ingestEndpointTransformationOut");
const ingestEndpointTransformationPatch_1 = require("../models/ingestEndpointTransformationPatch");
const ingestEndpointUpdate_1 = require("../models/ingestEndpointUpdate");
const listResponseIngestEndpointOut_1 = require("../models/listResponseIngestEndpointOut");
const request_1 = require("../request");
class IngestEndpoint {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    list(sourceId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint");
            request.setPathParam("source_id", sourceId);
            request.setQueryParams({
                limit: options === null || options === void 0 ? void 0 : options.limit,
                iterator: options === null || options === void 0 ? void 0 : options.iterator,
                order: options === null || options === void 0 ? void 0 : options.order,
            });
            return yield request.send(this.requestCtx, listResponseIngestEndpointOut_1.ListResponseIngestEndpointOutSerializer._fromJsonObject);
        });
    }
    create(sourceId, ingestEndpointIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source/{source_id}/endpoint");
            request.setPathParam("source_id", sourceId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(ingestEndpointIn_1.IngestEndpointInSerializer._toJsonObject(ingestEndpointIn));
            return yield request.send(this.requestCtx, ingestEndpointOut_1.IngestEndpointOutSerializer._fromJsonObject);
        });
    }
    get(sourceId, endpointId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}");
            request.setPathParam("source_id", sourceId);
            request.setPathParam("endpoint_id", endpointId);
            return yield request.send(this.requestCtx, ingestEndpointOut_1.IngestEndpointOutSerializer._fromJsonObject);
        });
    }
    update(sourceId, endpointId, ingestEndpointUpdate) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}");
            request.setPathParam("source_id", sourceId);
            request.setPathParam("endpoint_id", endpointId);
            request.setBody(ingestEndpointUpdate_1.IngestEndpointUpdateSerializer._toJsonObject(ingestEndpointUpdate));
            return yield request.send(this.requestCtx, ingestEndpointOut_1.IngestEndpointOutSerializer._fromJsonObject);
        });
    }
    delete(sourceId, endpointId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}");
            request.setPathParam("source_id", sourceId);
            request.setPathParam("endpoint_id", endpointId);
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    getHeaders(sourceId, endpointId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/headers");
            request.setPathParam("source_id", sourceId);
            request.setPathParam("endpoint_id", endpointId);
            return yield request.send(this.requestCtx, ingestEndpointHeadersOut_1.IngestEndpointHeadersOutSerializer._fromJsonObject);
        });
    }
    updateHeaders(sourceId, endpointId, ingestEndpointHeadersIn) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/headers");
            request.setPathParam("source_id", sourceId);
            request.setPathParam("endpoint_id", endpointId);
            request.setBody(ingestEndpointHeadersIn_1.IngestEndpointHeadersInSerializer._toJsonObject(ingestEndpointHeadersIn));
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    getSecret(sourceId, endpointId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/secret");
            request.setPathParam("source_id", sourceId);
            request.setPathParam("endpoint_id", endpointId);
            return yield request.send(this.requestCtx, ingestEndpointSecretOut_1.IngestEndpointSecretOutSerializer._fromJsonObject);
        });
    }
    rotateSecret(sourceId, endpointId, ingestEndpointSecretIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/secret/rotate");
            request.setPathParam("source_id", sourceId);
            request.setPathParam("endpoint_id", endpointId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(ingestEndpointSecretIn_1.IngestEndpointSecretInSerializer._toJsonObject(ingestEndpointSecretIn));
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    getTransformation(sourceId, endpointId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/transformation");
            request.setPathParam("source_id", sourceId);
            request.setPathParam("endpoint_id", endpointId);
            return yield request.send(this.requestCtx, ingestEndpointTransformationOut_1.IngestEndpointTransformationOutSerializer._fromJsonObject);
        });
    }
    setTransformation(sourceId, endpointId, ingestEndpointTransformationPatch) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/ingest/api/v1/source/{source_id}/endpoint/{endpoint_id}/transformation");
            request.setPathParam("source_id", sourceId);
            request.setPathParam("endpoint_id", endpointId);
            request.setBody(ingestEndpointTransformationPatch_1.IngestEndpointTransformationPatchSerializer._toJsonObject(ingestEndpointTransformationPatch));
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
}
exports.IngestEndpoint = IngestEndpoint;
