// svix/dist/api/ingestSource.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IngestSource = void 0;
const ingestSourceIn_1 = require("../models/ingestSourceIn");
const ingestSourceOut_1 = require("../models/ingestSourceOut");
const listResponseIngestSourceOut_1 = require("../models/listResponseIngestSourceOut");
const rotateTokenOut_1 = require("../models/rotateTokenOut");
const request_1 = require("../request");
class IngestSource {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    list(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source");
            request.setQueryParams({
                limit: options === null || options === void 0 ? void 0 : options.limit,
                iterator: options === null || options === void 0 ? void 0 : options.iterator,
                order: options === null || options === void 0 ? void 0 : options.order,
            });
            return yield request.send(this.requestCtx, listResponseIngestSourceOut_1.ListResponseIngestSourceOutSerializer._fromJsonObject);
        });
    }
    create(ingestSourceIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source");
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(ingestSourceIn_1.IngestSourceInSerializer._toJsonObject(ingestSourceIn));
            return yield request.send(this.requestCtx, ingestSourceOut_1.IngestSourceOutSerializer._fromJsonObject);
        });
    }
    get(sourceId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/ingest/api/v1/source/{source_id}");
            request.setPathParam("source_id", sourceId);
            return yield request.send(this.requestCtx, ingestSourceOut_1.IngestSourceOutSerializer._fromJsonObject);
        });
    }
    update(sourceId, ingestSourceIn) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/ingest/api/v1/source/{source_id}");
            request.setPathParam("source_id", sourceId);
            request.setBody(ingestSourceIn_1.IngestSourceInSerializer._toJsonObject(ingestSourceIn));
            return yield request.send(this.requestCtx, ingestSourceOut_1.IngestSourceOutSerializer._fromJsonObject);
        });
    }
    delete(sourceId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/ingest/api/v1/source/{source_id}");
            request.setPathParam("source_id", sourceId);
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    rotateToken(sourceId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/ingest/api/v1/source/{source_id}/token/rotate");
            request.setPathParam("source_id", sourceId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            return yield request.send(this.requestCtx, rotateTokenOut_1.RotateTokenOutSerializer._fromJsonObject);
        });
    }
}
exports.IngestSource = IngestSource;
