// svix/dist/api/integration.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Integration = void 0;
const integrationIn_1 = require("../models/integrationIn");
const integrationKeyOut_1 = require("../models/integrationKeyOut");
const integrationOut_1 = require("../models/integrationOut");
const integrationUpdate_1 = require("../models/integrationUpdate");
const listResponseIntegrationOut_1 = require("../models/listResponseIntegrationOut");
const request_1 = require("../request");
class Integration {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    list(appId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/integration");
            request.setPathParam("app_id", appId);
            request.setQueryParams({
                limit: options === null || options === void 0 ? void 0 : options.limit,
                iterator: options === null || options === void 0 ? void 0 : options.iterator,
                order: options === null || options === void 0 ? void 0 : options.order,
            });
            return yield request.send(this.requestCtx, listResponseIntegrationOut_1.ListResponseIntegrationOutSerializer._fromJsonObject);
        });
    }
    create(appId, integrationIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/integration");
            request.setPathParam("app_id", appId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(integrationIn_1.IntegrationInSerializer._toJsonObject(integrationIn));
            return yield request.send(this.requestCtx, integrationOut_1.IntegrationOutSerializer._fromJsonObject);
        });
    }
    get(appId, integId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/integration/{integ_id}");
            request.setPathParam("app_id", appId);
            request.setPathParam("integ_id", integId);
            return yield request.send(this.requestCtx, integrationOut_1.IntegrationOutSerializer._fromJsonObject);
        });
    }
    update(appId, integId, integrationUpdate) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/app/{app_id}/integration/{integ_id}");
            request.setPathParam("app_id", appId);
            request.setPathParam("integ_id", integId);
            request.setBody(integrationUpdate_1.IntegrationUpdateSerializer._toJsonObject(integrationUpdate));
            return yield request.send(this.requestCtx, integrationOut_1.IntegrationOutSerializer._fromJsonObject);
        });
    }
    delete(appId, integId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}/integration/{integ_id}");
            request.setPathParam("app_id", appId);
            request.setPathParam("integ_id", integId);
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    getKey(appId, integId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/integration/{integ_id}/key");
            request.setPathParam("app_id", appId);
            request.setPathParam("integ_id", integId);
            return yield request.send(this.requestCtx, integrationKeyOut_1.IntegrationKeyOutSerializer._fromJsonObject);
        });
    }
    rotateKey(appId, integId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/integration/{integ_id}/key/rotate");
            request.setPathParam("app_id", appId);
            request.setPathParam("integ_id", integId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            return yield request.send(this.requestCtx, integrationKeyOut_1.IntegrationKeyOutSerializer._fromJsonObject);
        });
    }
}
exports.Integration = Integration;
