// svix/dist/api/messagePoller.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessagePoller = void 0;
const pollingEndpointConsumerSeekIn_1 = require("../models/pollingEndpointConsumerSeekIn");
const pollingEndpointConsumerSeekOut_1 = require("../models/pollingEndpointConsumerSeekOut");
const pollingEndpointOut_1 = require("../models/pollingEndpointOut");
const request_1 = require("../request");
class MessagePoller {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    poll(appId, sinkId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/poller/{sink_id}");
            request.setPathParam("app_id", appId);
            request.setPathParam("sink_id", sinkId);
            request.setQueryParams({
                limit: options === null || options === void 0 ? void 0 : options.limit,
                iterator: options === null || options === void 0 ? void 0 : options.iterator,
                event_type: options === null || options === void 0 ? void 0 : options.eventType,
                channel: options === null || options === void 0 ? void 0 : options.channel,
                after: options === null || options === void 0 ? void 0 : options.after,
            });
            return yield request.send(this.requestCtx, pollingEndpointOut_1.PollingEndpointOutSerializer._fromJsonObject);
        });
    }
    consumerPoll(appId, sinkId, consumerId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/poller/{sink_id}/consumer/{consumer_id}");
            request.setPathParam("app_id", appId);
            request.setPathParam("sink_id", sinkId);
            request.setPathParam("consumer_id", consumerId);
            request.setQueryParams({
                limit: options === null || options === void 0 ? void 0 : options.limit,
                iterator: options === null || options === void 0 ? void 0 : options.iterator,
            });
            return yield request.send(this.requestCtx, pollingEndpointOut_1.PollingEndpointOutSerializer._fromJsonObject);
        });
    }
    consumerSeek(appId, sinkId, consumerId, pollingEndpointConsumerSeekIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/poller/{sink_id}/consumer/{consumer_id}/seek");
            request.setPathParam("app_id", appId);
            request.setPathParam("sink_id", sinkId);
            request.setPathParam("consumer_id", consumerId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(pollingEndpointConsumerSeekIn_1.PollingEndpointConsumerSeekInSerializer._toJsonObject(pollingEndpointConsumerSeekIn));
            return yield request.send(this.requestCtx, pollingEndpointConsumerSeekOut_1.PollingEndpointConsumerSeekOutSerializer._fromJsonObject);
        });
    }
}
exports.MessagePoller = MessagePoller;
