// svix/dist/api/message.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.messageInRaw = exports.Message = void 0;
const expungeAllContentsOut_1 = require("../models/expungeAllContentsOut");
const listResponseMessageOut_1 = require("../models/listResponseMessageOut");
const messageOut_1 = require("../models/messageOut");
const messagePrecheckIn_1 = require("../models/messagePrecheckIn");
const messagePrecheckOut_1 = require("../models/messagePrecheckOut");
const messagePoller_1 = require("./messagePoller");
const request_1 = require("../request");
const messageIn_1 = require("../models/messageIn");
class Message {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    get poller() {
        return new messagePoller_1.MessagePoller(this.requestCtx);
    }
    list(appId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/msg");
            request.setPathParam("app_id", appId);
            request.setQueryParams({
                limit: options === null || options === void 0 ? void 0 : options.limit,
                iterator: options === null || options === void 0 ? void 0 : options.iterator,
                channel: options === null || options === void 0 ? void 0 : options.channel,
                before: options === null || options === void 0 ? void 0 : options.before,
                after: options === null || options === void 0 ? void 0 : options.after,
                with_content: options === null || options === void 0 ? void 0 : options.withContent,
                tag: options === null || options === void 0 ? void 0 : options.tag,
                event_types: options === null || options === void 0 ? void 0 : options.eventTypes,
            });
            return yield request.send(this.requestCtx, listResponseMessageOut_1.ListResponseMessageOutSerializer._fromJsonObject);
        });
    }
    create(appId, messageIn, options) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/msg");
            request.setPathParam("app_id", appId);
            request.setQueryParams({
                with_content: false,
            });
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(messageIn_1.MessageInSerializer._toJsonObject(messageIn));
            const response = yield request.send(this.requestCtx, messageOut_1.MessageOutSerializer._fromJsonObject);
            if ((_a = options === null || options === void 0 ? void 0 : options.withContent) !== null && _a !== void 0 ? _a : true) {
                response.payload = messageIn.payload;
            }
            return response;
        });
    }
    expungeAllContents(appId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/msg/expunge-all-contents");
            request.setPathParam("app_id", appId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            return yield request.send(this.requestCtx, expungeAllContentsOut_1.ExpungeAllContentsOutSerializer._fromJsonObject);
        });
    }
    precheck(appId, messagePrecheckIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/msg/precheck/active");
            request.setPathParam("app_id", appId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(messagePrecheckIn_1.MessagePrecheckInSerializer._toJsonObject(messagePrecheckIn));
            return yield request.send(this.requestCtx, messagePrecheckOut_1.MessagePrecheckOutSerializer._fromJsonObject);
        });
    }
    get(appId, msgId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/msg/{msg_id}");
            request.setPathParam("app_id", appId);
            request.setPathParam("msg_id", msgId);
            request.setQueryParams({
                with_content: options === null || options === void 0 ? void 0 : options.withContent,
            });
            return yield request.send(this.requestCtx, messageOut_1.MessageOutSerializer._fromJsonObject);
        });
    }
    expungeContent(appId, msgId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/app/{app_id}/msg/{msg_id}/content");
            request.setPathParam("app_id", appId);
            request.setPathParam("msg_id", msgId);
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
}
exports.Message = Message;
function messageInRaw(eventType, payload, contentType) {
    const headers = contentType ? { "content-type": contentType } : undefined;
    return {
        eventType,
        payload: {},
        transformationsParams: {
            rawPayload: payload,
            headers,
        },
    };
}
exports.messageInRaw = messageInRaw;
