// svix/dist/api/streaming.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Streaming = void 0;
const endpointHeadersOut_1 = require("../models/endpointHeadersOut");
const httpSinkHeadersPatchIn_1 = require("../models/httpSinkHeadersPatchIn");
const sinkTransformationOut_1 = require("../models/sinkTransformationOut");
const streamingEventType_1 = require("./streamingEventType");
const streamingEvents_1 = require("./streamingEvents");
const streamingSink_1 = require("./streamingSink");
const streamingStream_1 = require("./streamingStream");
const request_1 = require("../request");
class Streaming {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    get event_type() {
        return new streamingEventType_1.StreamingEventType(this.requestCtx);
    }
    get events() {
        return new streamingEvents_1.StreamingEvents(this.requestCtx);
    }
    get sink() {
        return new streamingSink_1.StreamingSink(this.requestCtx);
    }
    get stream() {
        return new streamingStream_1.StreamingStream(this.requestCtx);
    }
    sinkHeadersGet(streamId, sinkId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink/{sink_id}/headers");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            return yield request.send(this.requestCtx, endpointHeadersOut_1.EndpointHeadersOutSerializer._fromJsonObject);
        });
    }
    sinkHeadersPatch(streamId, sinkId, httpSinkHeadersPatchIn) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/stream/{stream_id}/sink/{sink_id}/headers");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            request.setBody(httpSinkHeadersPatchIn_1.HttpSinkHeadersPatchInSerializer._toJsonObject(httpSinkHeadersPatchIn));
            return yield request.send(this.requestCtx, endpointHeadersOut_1.EndpointHeadersOutSerializer._fromJsonObject);
        });
    }
    sinkTransformationGet(streamId, sinkId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink/{sink_id}/transformation");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            return yield request.send(this.requestCtx, sinkTransformationOut_1.SinkTransformationOutSerializer._fromJsonObject);
        });
    }
}
exports.Streaming = Streaming;
