// svix/dist/api/streamingSink.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamingSink = void 0;
const emptyResponse_1 = require("../models/emptyResponse");
const endpointSecretRotateIn_1 = require("../models/endpointSecretRotateIn");
const listResponseStreamSinkOut_1 = require("../models/listResponseStreamSinkOut");
const sinkSecretOut_1 = require("../models/sinkSecretOut");
const sinkTransformIn_1 = require("../models/sinkTransformIn");
const streamSinkIn_1 = require("../models/streamSinkIn");
const streamSinkOut_1 = require("../models/streamSinkOut");
const streamSinkPatch_1 = require("../models/streamSinkPatch");
const request_1 = require("../request");
class StreamingSink {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    list(streamId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink");
            request.setPathParam("stream_id", streamId);
            request.setQueryParams({
                limit: options === null || options === void 0 ? void 0 : options.limit,
                iterator: options === null || options === void 0 ? void 0 : options.iterator,
                order: options === null || options === void 0 ? void 0 : options.order,
            });
            return yield request.send(this.requestCtx, listResponseStreamSinkOut_1.ListResponseStreamSinkOutSerializer._fromJsonObject);
        });
    }
    create(streamId, streamSinkIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream/{stream_id}/sink");
            request.setPathParam("stream_id", streamId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(streamSinkIn_1.StreamSinkInSerializer._toJsonObject(streamSinkIn));
            return yield request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
        });
    }
    get(streamId, sinkId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink/{sink_id}");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            return yield request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
        });
    }
    update(streamId, sinkId, streamSinkIn) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PUT, "/api/v1/stream/{stream_id}/sink/{sink_id}");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            request.setBody(streamSinkIn_1.StreamSinkInSerializer._toJsonObject(streamSinkIn));
            return yield request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
        });
    }
    delete(streamId, sinkId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.DELETE, "/api/v1/stream/{stream_id}/sink/{sink_id}");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
    patch(streamId, sinkId, streamSinkPatch) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/stream/{stream_id}/sink/{sink_id}");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            request.setBody(streamSinkPatch_1.StreamSinkPatchSerializer._toJsonObject(streamSinkPatch));
            return yield request.send(this.requestCtx, streamSinkOut_1.StreamSinkOutSerializer._fromJsonObject);
        });
    }
    getSecret(streamId, sinkId) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/stream/{stream_id}/sink/{sink_id}/secret");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            return yield request.send(this.requestCtx, sinkSecretOut_1.SinkSecretOutSerializer._fromJsonObject);
        });
    }
    rotateSecret(streamId, sinkId, endpointSecretRotateIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/stream/{stream_id}/sink/{sink_id}/secret/rotate");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(endpointSecretRotateIn_1.EndpointSecretRotateInSerializer._toJsonObject(endpointSecretRotateIn));
            return yield request.send(this.requestCtx, emptyResponse_1.EmptyResponseSerializer._fromJsonObject);
        });
    }
    transformationPartialUpdate(streamId, sinkId, sinkTransformIn) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.PATCH, "/api/v1/stream/{stream_id}/sink/{sink_id}/transformation");
            request.setPathParam("stream_id", streamId);
            request.setPathParam("sink_id", sinkId);
            request.setBody(sinkTransformIn_1.SinkTransformInSerializer._toJsonObject(sinkTransformIn));
            return yield request.send(this.requestCtx, emptyResponse_1.EmptyResponseSerializer._fromJsonObject);
        });
    }
}
exports.StreamingSink = StreamingSink;
