// dodopayments/src/resources/product-collections/product-collections.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../core/resource';
import * as GroupsAPI from './groups/groups';
import {
  GroupCreateParams,
  GroupDeleteParams,
  GroupProduct,
  GroupUpdateParams,
  Groups,
  ProductCollectionGroupDetails,
  ProductCollectionGroupResponse,
} from './groups/groups';
import { APIPromise } from '../../core/api-promise';
import {
  DefaultPageNumberPagination,
  type DefaultPageNumberPaginationParams,
  PagePromise,
} from '../../core/pagination';
import { buildHeaders } from '../../internal/headers';
import { RequestOptions } from '../../internal/request-options';
import { path } from '../../internal/utils/path';

export class ProductCollections extends APIResource {
  groups: GroupsAPI.Groups = new GroupsAPI.Groups(this._client);

  /**
   * @example
   * ```ts
   * // Automatically fetches more pages as needed.
   * for await (const productCollectionListResponse of client.productCollections.list()) {
   *   // ...
   * }
   * ```
   */
  list(
    query: ProductCollectionListParams | null | undefined = {},
    options?: RequestOptions,
  ): PagePromise<ProductCollectionListResponsesDefaultPageNumberPagination, ProductCollectionListResponse> {
    return this._client.getAPIList(
      '/product-collections',
      DefaultPageNumberPagination<ProductCollectionListResponse>,
      { query, ...options },
    );
  }

  /**
   * @example
   * ```ts
   * const productCollection =
   *   await client.productCollections.create({
   *     groups: [{ products: [{ product_id: 'product_id' }] }],
   *     name: 'name',
   *   });
   * ```
   */
  create(body: ProductCollectionCreateParams, options?: RequestOptions): APIPromise<ProductCollection> {
    return this._client.post('/product-collections', { body, ...options });
  }

  /**
   * @example
   * ```ts
   * const productCollection =
   *   await client.productCollections.retrieve(
   *     'pdc_8BWv0hojwUH7iCDabr0NI',
   *   );
   * ```
   */
  retrieve(id: string, options?: RequestOptions): APIPromise<ProductCollection> {
    return this._client.get(path`/product-collections/${id}`, options);
  }

  /**
   * @example
   * ```ts
   * await client.productCollections.delete(
   *   'pdc_8BWv0hojwUH7iCDabr0NI',
   * );
   * ```
   */
  delete(id: string, options?: RequestOptions): APIPromise<void> {
    return this._client.delete(path`/product-collections/${id}`, {
      ...options,
      headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
    });
  }

  /**
   * @example
   * ```ts
   * await client.productCollections.update(
   *   'pdc_8BWv0hojwUH7iCDabr0NI',
   * );
   * ```
   */
  update(id: string, body: ProductCollectionUpdateParams, options?: RequestOptions): APIPromise<void> {
    return this._client.patch(path`/product-collections/${id}`, {
      body,
      ...options,
      headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
    });
  }

  /**
   * @example
   * ```ts
   * const response =
   *   await client.productCollections.updateImages(
   *     'pdc_8BWv0hojwUH7iCDabr0NI',
   *   );
   * ```
   */
  updateImages(
    id: string,
    params: ProductCollectionUpdateImagesParams | null | undefined = {},
    options?: RequestOptions,
  ): APIPromise<ProductCollectionUpdateImagesResponse> {
    const { force_update } = params ?? {};
    return this._client.put(path`/product-collections/${id}/images`, { query: { force_update }, ...options });
  }

  /**
   * @example
   * ```ts
   * const response = await client.productCollections.unarchive(
   *   'pdc_8BWv0hojwUH7iCDabr0NI',
   * );
   * ```
   */
  unarchive(id: string, options?: RequestOptions): APIPromise<ProductCollectionUnarchiveResponse> {
    return this._client.post(path`/product-collections/${id}/unarchive`, options);
  }
}

export type ProductCollectionListResponsesDefaultPageNumberPagination =
  DefaultPageNumberPagination<ProductCollectionListResponse>;

export interface ProductCollection {
  /**
   * Unique identifier for the product collection
   */
  id: string;

  /**
   * Brand ID for the collection
   */
  brand_id: string;

  /**
   * Timestamp when the collection was created
   */
  created_at: string;

  /**
   * Groups in this collection
   */
  groups: Array<GroupsAPI.ProductCollectionGroupResponse>;

  /**
   * Name of the collection
   */
  name: string;

  /**
   * Timestamp when the collection was last updated
   */
  updated_at: string;

  /**
   * Description of the collection
   */
  description?: string | null;

  /**
   * Default effective_at setting for subscription plan downgrades (null = inherit
   * from business)
   */
  effective_at_on_downgrade?: 'immediately' | 'next_billing_date' | null;

  /**
   * Default effective_at setting for subscription plan upgrades (null = inherit from
   * business)
   */
  effective_at_on_upgrade?: 'immediately' | 'next_billing_date' | null;

  /**
   * URL of the collection image
   */
  image?: string | null;

  /**
   * Default behavior for subscription plan changes on payment failure (null =
   * inherit from business)
   */
  on_payment_failure?: 'prevent_change' | 'apply_change' | null;

  /**
   * Default proration billing mode for subscription plan downgrades (null = inherit
   * from business)
   */
  proration_billing_mode_on_downgrade?:
    | 'prorated_immediately'
    | 'full_immediately'
    | 'difference_immediately'
    | 'do_not_bill'
    | null;

  /**
   * Default proration billing mode for subscription plan upgrades (null = inherit
   * from business)
   */
  proration_billing_mode_on_upgrade?:
    | 'prorated_immediately'
    | 'full_immediately'
    | 'difference_immediately'
    | 'do_not_bill'
    | null;
}

export interface ProductCollectionListResponse {
  /**
   * Collection ID
   */
  id: string;

  /**
   * Timestamp when created
   */
  created_at: string;

  /**
   * Collection name
   */
  name: string;

  /**
   * Number of products in the collection
   */
  products_count: number;

  /**
   * Timestamp when last updated
   */
  updated_at: string;

  /**
   * Collection description
   */
  description?: string | null;

  /**
   * Collection image URL
   */
  image?: string | null;
}

export interface ProductCollectionUnarchiveResponse {
  /**
   * Collection ID that was unarchived
   */
  collection_id: string;

  /**
   * Product IDs that were excluded because they are archived
   */
  excluded_product_ids: Array<string>;

  /**
   * Success message
   */
  message: string;
}

export interface ProductCollectionUpdateImagesResponse {
  /**
   * Presigned S3 URL for uploading the image
   */
  url: string;

  /**
   * Optional image ID (present when force_update is true)
   */
  image_id?: string | null;
}

export interface ProductCollectionListParams extends DefaultPageNumberPaginationParams {
  /**
   * List archived collections
   */
  archived?: boolean;

  /**
   * Filter by Brand id
   */
  brand_id?: string;
}

export interface ProductCollectionCreateParams {
  /**
   * Groups of products in this collection
   */
  groups: Array<GroupsAPI.ProductCollectionGroupDetails>;

  /**
   * Name of the product collection
   */
  name: string;

  /**
   * Brand id for the collection, if not provided will default to primary brand
   */
  brand_id?: string | null;

  /**
   * Optional description of the product collection
   */
  description?: string | null;

  /**
   * Default effective_at setting for subscription plan downgrades (NULL = inherit
   * from business)
   */
  effective_at_on_downgrade?: 'immediately' | 'next_billing_date' | null;

  /**
   * Default effective_at setting for subscription plan upgrades (NULL = inherit from
   * business)
   */
  effective_at_on_upgrade?: 'immediately' | 'next_billing_date' | null;

  /**
   * Default behavior for subscription plan changes on payment failure (NULL =
   * inherit from business)
   */
  on_payment_failure?: 'prevent_change' | 'apply_change' | null;

  /**
   * Default proration billing mode for subscription plan downgrades (NULL = inherit
   * from business)
   */
  proration_billing_mode_on_downgrade?:
    | 'prorated_immediately'
    | 'full_immediately'
    | 'difference_immediately'
    | 'do_not_bill'
    | null;

  /**
   * Default proration billing mode for subscription plan upgrades (NULL = inherit
   * from business)
   */
  proration_billing_mode_on_upgrade?:
    | 'prorated_immediately'
    | 'full_immediately'
    | 'difference_immediately'
    | 'do_not_bill'
    | null;
}

export interface ProductCollectionUpdateParams {
  /**
   * Optional brand_id update
   */
  brand_id?: string | null;

  /**
   * Optional description update - pass null to remove, omit to keep unchanged
   */
  description?: string | null;

  /**
   * Effective_at setting for downgrades: Some(Some(val)) = set, Some(None) = clear
   * (inherit), None = no change
   */
  effective_at_on_downgrade?: 'immediately' | 'next_billing_date' | null;

  /**
   * Effective_at setting for upgrades: Some(Some(val)) = set, Some(None) = clear
   * (inherit), None = no change
   */
  effective_at_on_upgrade?: 'immediately' | 'next_billing_date' | null;

  /**
   * Optional new order for groups (array of group UUIDs in desired order)
   */
  group_order?: Array<string> | null;

  /**
   * Optional image update - pass null to remove, omit to keep unchanged
   */
  image_id?: string | null;

  /**
   * Optional new name for the collection
   */
  name?: string | null;

  /**
   * On payment failure behavior: Some(Some(val)) = set, Some(None) = clear
   * (inherit), None = no change
   */
  on_payment_failure?: 'prevent_change' | 'apply_change' | null;

  /**
   * Proration billing mode for downgrades: Some(Some(val)) = set, Some(None) = clear
   * (inherit), None = no change
   */
  proration_billing_mode_on_downgrade?:
    | 'prorated_immediately'
    | 'full_immediately'
    | 'difference_immediately'
    | 'do_not_bill'
    | null;

  /**
   * Proration billing mode for upgrades: Some(Some(val)) = set, Some(None) = clear
   * (inherit), None = no change
   */
  proration_billing_mode_on_upgrade?:
    | 'prorated_immediately'
    | 'full_immediately'
    | 'difference_immediately'
    | 'do_not_bill'
    | null;
}

export interface ProductCollectionUpdateImagesParams {
  /**
   * If true, generates a new image ID to force cache invalidation
   */
  force_update?: boolean | null;
}

ProductCollections.Groups = Groups;

export declare namespace ProductCollections {
  export {
    type ProductCollection as ProductCollection,
    type ProductCollectionListResponse as ProductCollectionListResponse,
    type ProductCollectionUnarchiveResponse as ProductCollectionUnarchiveResponse,
    type ProductCollectionUpdateImagesResponse as ProductCollectionUpdateImagesResponse,
    type ProductCollectionListResponsesDefaultPageNumberPagination as ProductCollectionListResponsesDefaultPageNumberPagination,
    type ProductCollectionListParams as ProductCollectionListParams,
    type ProductCollectionCreateParams as ProductCollectionCreateParams,
    type ProductCollectionUpdateParams as ProductCollectionUpdateParams,
    type ProductCollectionUpdateImagesParams as ProductCollectionUpdateImagesParams,
  };

  export {
    Groups as Groups,
    type GroupProduct as GroupProduct,
    type ProductCollectionGroupDetails as ProductCollectionGroupDetails,
    type ProductCollectionGroupResponse as ProductCollectionGroupResponse,
    type GroupCreateParams as GroupCreateParams,
    type GroupDeleteParams as GroupDeleteParams,
    type GroupUpdateParams as GroupUpdateParams,
  };
}
