// dodopayments/resources/products/localized-prices.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalizedPrices = void 0;
const resource_1 = require("../../core/resource.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class LocalizedPrices extends resource_1.APIResource {
    /**
     * @example
     * ```ts
     * const listLocalizedPricesResponse =
     *   await client.products.localizedPrices.list(
     *     'pdt_R8AWMPiV8RyJElcCKvAID',
     *   );
     * ```
     */
    list(productID, options) {
        return this._client.get((0, path_1.path) `/products/${productID}/localized-prices`, options);
    }
    /**
     * @example
     * ```ts
     * const localizedPrice =
     *   await client.products.localizedPrices.create(
     *     'pdt_R8AWMPiV8RyJElcCKvAID',
     *     { amount: 0, currency: 'AED' },
     *   );
     * ```
     */
    create(productID, body, options) {
        return this._client.post((0, path_1.path) `/products/${productID}/localized-prices`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const localizedPrice =
     *   await client.products.localizedPrices.retrieve(
     *     'lcp_3aOOT7ebrzBOV41yL2V6s',
     *     { product_id: 'pdt_R8AWMPiV8RyJElcCKvAID' },
     *   );
     * ```
     */
    retrieve(id, params, options) {
        const { product_id } = params;
        return this._client.get((0, path_1.path) `/products/${product_id}/localized-prices/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * const localizedPrice =
     *   await client.products.localizedPrices.update(
     *     'lcp_3aOOT7ebrzBOV41yL2V6s',
     *     { product_id: 'pdt_R8AWMPiV8RyJElcCKvAID' },
     *   );
     * ```
     */
    update(id, params, options) {
        const { product_id, ...body } = params;
        return this._client.patch((0, path_1.path) `/products/${product_id}/localized-prices/${id}`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * await client.products.localizedPrices.archive(
     *   'lcp_3aOOT7ebrzBOV41yL2V6s',
     *   { product_id: 'pdt_R8AWMPiV8RyJElcCKvAID' },
     * );
     * ```
     */
    archive(id, params, options) {
        const { product_id } = params;
        return this._client.delete((0, path_1.path) `/products/${product_id}/localized-prices/${id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
exports.LocalizedPrices = LocalizedPrices;
