// dodopayments/resources/products/localized-prices.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import { buildHeaders } from "../../internal/headers.mjs";
import { path } from "../../internal/utils/path.mjs";
export class LocalizedPrices extends APIResource {
    /**
     * @example
     * ```ts
     * const listLocalizedPricesResponse =
     *   await client.products.localizedPrices.list(
     *     'pdt_R8AWMPiV8RyJElcCKvAID',
     *   );
     * ```
     */
    list(productID, options) {
        return this._client.get(path `/products/${productID}/localized-prices`, options);
    }
    /**
     * @example
     * ```ts
     * const localizedPrice =
     *   await client.products.localizedPrices.create(
     *     'pdt_R8AWMPiV8RyJElcCKvAID',
     *     { amount: 0, currency: 'AED' },
     *   );
     * ```
     */
    create(productID, body, options) {
        return this._client.post(path `/products/${productID}/localized-prices`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const localizedPrice =
     *   await client.products.localizedPrices.retrieve(
     *     'lcp_3aOOT7ebrzBOV41yL2V6s',
     *     { product_id: 'pdt_R8AWMPiV8RyJElcCKvAID' },
     *   );
     * ```
     */
    retrieve(id, params, options) {
        const { product_id } = params;
        return this._client.get(path `/products/${product_id}/localized-prices/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * const localizedPrice =
     *   await client.products.localizedPrices.update(
     *     'lcp_3aOOT7ebrzBOV41yL2V6s',
     *     { product_id: 'pdt_R8AWMPiV8RyJElcCKvAID' },
     *   );
     * ```
     */
    update(id, params, options) {
        const { product_id, ...body } = params;
        return this._client.patch(path `/products/${product_id}/localized-prices/${id}`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * await client.products.localizedPrices.archive(
     *   'lcp_3aOOT7ebrzBOV41yL2V6s',
     *   { product_id: 'pdt_R8AWMPiV8RyJElcCKvAID' },
     * );
     * ```
     */
    archive(id, params, options) {
        const { product_id } = params;
        return this._client.delete(path `/products/${product_id}/localized-prices/${id}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
