// dodopayments/resources/products/products.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as ImagesAPI from "./images.mjs";
import { Images } from "./images.mjs";
import * as LocalizedPricesAPI from "./localized-prices.mjs";
import { LocalizedPrices, } from "./localized-prices.mjs";
import * as ShortLinksAPI from "./short-links.mjs";
import { ShortLinks, } from "./short-links.mjs";
import { DefaultPageNumberPagination, } from "../../core/pagination.mjs";
import { buildHeaders } from "../../internal/headers.mjs";
import { path } from "../../internal/utils/path.mjs";
export class Products extends APIResource {
    constructor() {
        super(...arguments);
        this.images = new ImagesAPI.Images(this._client);
        this.shortLinks = new ShortLinksAPI.ShortLinks(this._client);
        this.localizedPrices = new LocalizedPricesAPI.LocalizedPrices(this._client);
    }
    /**
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const productListResponse of client.products.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/products', (DefaultPageNumberPagination), {
            query,
            ...options,
        });
    }
    /**
     * @example
     * ```ts
     * const product = await client.products.create({
     *   name: 'name',
     *   price: {
     *     currency: 'AED',
     *     discount: 0,
     *     price: 0,
     *     purchasing_power_parity: true,
     *     type: 'one_time_price',
     *   },
     *   tax_category: 'digital_products',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/products', { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const product = await client.products.retrieve(
     *   'pdt_R8AWMPiV8RyJElcCKvAID',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/products/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * await client.products.update('pdt_R8AWMPiV8RyJElcCKvAID');
     * ```
     */
    update(id, body, options) {
        return this._client.patch(path `/products/${id}`, {
            body,
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * await client.products.archive('pdt_R8AWMPiV8RyJElcCKvAID');
     * ```
     */
    archive(id, options) {
        return this._client.delete(path `/products/${id}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * await client.products.unarchive(
     *   'pdt_R8AWMPiV8RyJElcCKvAID',
     * );
     * ```
     */
    unarchive(id, options) {
        return this._client.post(path `/products/${id}/unarchive`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * const response = await client.products.updateFiles(
     *   'pdt_R8AWMPiV8RyJElcCKvAID',
     *   { file_name: 'file_name' },
     * );
     * ```
     */
    updateFiles(id, body, options) {
        return this._client.put(path `/products/${id}/files`, { body, ...options });
    }
}
Products.Images = Images;
Products.ShortLinks = ShortLinks;
Products.LocalizedPrices = LocalizedPrices;
