// dodopayments/resources/products/products.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Products = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const ImagesAPI = tslib_1.__importStar(require("./images.js"));
const images_1 = require("./images.js");
const LocalizedPricesAPI = tslib_1.__importStar(require("./localized-prices.js"));
const localized_prices_1 = require("./localized-prices.js");
const ShortLinksAPI = tslib_1.__importStar(require("./short-links.js"));
const short_links_1 = require("./short-links.js");
const pagination_1 = require("../../core/pagination.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class Products extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.images = new ImagesAPI.Images(this._client);
        this.shortLinks = new ShortLinksAPI.ShortLinks(this._client);
        this.localizedPrices = new LocalizedPricesAPI.LocalizedPrices(this._client);
    }
    /**
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const productListResponse of client.products.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/products', (pagination_1.DefaultPageNumberPagination), {
            query,
            ...options,
        });
    }
    /**
     * @example
     * ```ts
     * const product = await client.products.create({
     *   name: 'name',
     *   price: {
     *     currency: 'AED',
     *     discount: 0,
     *     price: 0,
     *     purchasing_power_parity: true,
     *     type: 'one_time_price',
     *   },
     *   tax_category: 'digital_products',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/products', { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const product = await client.products.retrieve(
     *   'pdt_R8AWMPiV8RyJElcCKvAID',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/products/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * await client.products.update('pdt_R8AWMPiV8RyJElcCKvAID');
     * ```
     */
    update(id, body, options) {
        return this._client.patch((0, path_1.path) `/products/${id}`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * await client.products.archive('pdt_R8AWMPiV8RyJElcCKvAID');
     * ```
     */
    archive(id, options) {
        return this._client.delete((0, path_1.path) `/products/${id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * await client.products.unarchive(
     *   'pdt_R8AWMPiV8RyJElcCKvAID',
     * );
     * ```
     */
    unarchive(id, options) {
        return this._client.post((0, path_1.path) `/products/${id}/unarchive`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * const response = await client.products.updateFiles(
     *   'pdt_R8AWMPiV8RyJElcCKvAID',
     *   { file_name: 'file_name' },
     * );
     * ```
     */
    updateFiles(id, body, options) {
        return this._client.put((0, path_1.path) `/products/${id}/files`, { body, ...options });
    }
}
exports.Products = Products;
Products.Images = images_1.Images;
Products.ShortLinks = short_links_1.ShortLinks;
Products.LocalizedPrices = localized_prices_1.LocalizedPrices;
