// @dodopayments/core/dist/webhook/webhook.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// ../../node_modules/@stablelib/base64/lib/base64.js
var require_base64 = __commonJS({
  "../../node_modules/@stablelib/base64/lib/base64.js"(exports2) {
    "use strict";
    var __extends = exports2 && exports2.__extends || /* @__PURE__ */ function() {
      var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf || { __proto__: [] } instanceof Array && function(d2, b2) {
          d2.__proto__ = b2;
        } || function(d2, b2) {
          for (var p in b2) if (b2.hasOwnProperty(p)) d2[p] = b2[p];
        };
        return extendStatics(d, b);
      };
      return function(d, b) {
        extendStatics(d, b);
        function __() {
          this.constructor = d;
        }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
      };
    }();
    Object.defineProperty(exports2, "__esModule", { value: true });
    var INVALID_BYTE = 256;
    var Coder = (
      /** @class */
      function() {
        function Coder2(_paddingCharacter) {
          if (_paddingCharacter === void 0) {
            _paddingCharacter = "=";
          }
          this._paddingCharacter = _paddingCharacter;
        }
        Coder2.prototype.encodedLength = function(length) {
          if (!this._paddingCharacter) {
            return (length * 8 + 5) / 6 | 0;
          }
          return (length + 2) / 3 * 4 | 0;
        };
        Coder2.prototype.encode = function(data) {
          var out = "";
          var i = 0;
          for (; i < data.length - 2; i += 3) {
            var c = data[i] << 16 | data[i + 1] << 8 | data[i + 2];
            out += this._encodeByte(c >>> 3 * 6 & 63);
            out += this._encodeByte(c >>> 2 * 6 & 63);
            out += this._encodeByte(c >>> 1 * 6 & 63);
            out += this._encodeByte(c >>> 0 * 6 & 63);
          }
          var left = data.length - i;
          if (left > 0) {
            var c = data[i] << 16 | (left === 2 ? data[i + 1] << 8 : 0);
            out += this._encodeByte(c >>> 3 * 6 & 63);
            out += this._encodeByte(c >>> 2 * 6 & 63);
            if (left === 2) {
              out += this._encodeByte(c >>> 1 * 6 & 63);
            } else {
              out += this._paddingCharacter || "";
            }
            out += this._paddingCharacter || "";
          }
          return out;
        };
        Coder2.prototype.maxDecodedLength = function(length) {
          if (!this._paddingCharacter) {
            return (length * 6 + 7) / 8 | 0;
          }
          return length / 4 * 3 | 0;
        };
        Coder2.prototype.decodedLength = function(s) {
          return this.maxDecodedLength(s.length - this._getPaddingLength(s));
        };
        Coder2.prototype.decode = function(s) {
          if (s.length === 0) {
            return new Uint8Array(0);
          }
          var paddingLength = this._getPaddingLength(s);
          var length = s.length - paddingLength;
          var out = new Uint8Array(this.maxDecodedLength(length));
          var op = 0;
          var i = 0;
          var haveBad = 0;
          var v0 = 0, v1 = 0, v2 = 0, v3 = 0;
          for (; i < length - 4; i += 4) {
            v0 = this._decodeChar(s.charCodeAt(i + 0));
            v1 = this._decodeChar(s.charCodeAt(i + 1));
            v2 = this._decodeChar(s.charCodeAt(i + 2));
            v3 = this._decodeChar(s.charCodeAt(i + 3));
            out[op++] = v0 << 2 | v1 >>> 4;
            out[op++] = v1 << 4 | v2 >>> 2;
            out[op++] = v2 << 6 | v3;
            haveBad |= v0 & INVALID_BYTE;
            haveBad |= v1 & INVALID_BYTE;
            haveBad |= v2 & INVALID_BYTE;
            haveBad |= v3 & INVALID_BYTE;
          }
          if (i < length - 1) {
            v0 = this._decodeChar(s.charCodeAt(i));
            v1 = this._decodeChar(s.charCodeAt(i + 1));
            out[op++] = v0 << 2 | v1 >>> 4;
            haveBad |= v0 & INVALID_BYTE;
            haveBad |= v1 & INVALID_BYTE;
          }
          if (i < length - 2) {
            v2 = this._decodeChar(s.charCodeAt(i + 2));
            out[op++] = v1 << 4 | v2 >>> 2;
            haveBad |= v2 & INVALID_BYTE;
          }
          if (i < length - 3) {
            v3 = this._decodeChar(s.charCodeAt(i + 3));
            out[op++] = v2 << 6 | v3;
            haveBad |= v3 & INVALID_BYTE;
          }
          if (haveBad !== 0) {
            throw new Error("Base64Coder: incorrect characters for decoding");
          }
          return out;
        };
        Coder2.prototype._encodeByte = function(b) {
          var result = b;
          result += 65;
          result += 25 - b >>> 8 & 0 - 65 - 26 + 97;
          result += 51 - b >>> 8 & 26 - 97 - 52 + 48;
          result += 61 - b >>> 8 & 52 - 48 - 62 + 43;
          result += 62 - b >>> 8 & 62 - 43 - 63 + 47;
          return String.fromCharCode(result);
        };
        Coder2.prototype._decodeChar = function(c) {
          var result = INVALID_BYTE;
          result += (42 - c & c - 44) >>> 8 & -INVALID_BYTE + c - 43 + 62;
          result += (46 - c & c - 48) >>> 8 & -INVALID_BYTE + c - 47 + 63;
          result += (47 - c & c - 58) >>> 8 & -INVALID_BYTE + c - 48 + 52;
          result += (64 - c & c - 91) >>> 8 & -INVALID_BYTE + c - 65 + 0;
          result += (96 - c & c - 123) >>> 8 & -INVALID_BYTE + c - 97 + 26;
          return result;
        };
        Coder2.prototype._getPaddingLength = function(s) {
          var paddingLength = 0;
          if (this._paddingCharacter) {
            for (var i = s.length - 1; i >= 0; i--) {
              if (s[i] !== this._paddingCharacter) {
                break;
              }
              paddingLength++;
            }
            if (s.length < 4 || paddingLength > 2) {
              throw new Error("Base64Coder: incorrect padding");
            }
          }
          return paddingLength;
        };
        return Coder2;
      }()
    );
    exports2.Coder = Coder;
    var stdCoder = new Coder();
    function encode2(data) {
      return stdCoder.encode(data);
    }
    exports2.encode = encode2;
    function decode2(s) {
      return stdCoder.decode(s);
    }
    exports2.decode = decode2;
    var URLSafeCoder = (
      /** @class */
      function(_super) {
        __extends(URLSafeCoder2, _super);
        function URLSafeCoder2() {
          return _super !== null && _super.apply(this, arguments) || this;
        }
        URLSafeCoder2.prototype._encodeByte = function(b) {
          var result = b;
          result += 65;
          result += 25 - b >>> 8 & 0 - 65 - 26 + 97;
          result += 51 - b >>> 8 & 26 - 97 - 52 + 48;
          result += 61 - b >>> 8 & 52 - 48 - 62 + 45;
          result += 62 - b >>> 8 & 62 - 45 - 63 + 95;
          return String.fromCharCode(result);
        };
        URLSafeCoder2.prototype._decodeChar = function(c) {
          var result = INVALID_BYTE;
          result += (44 - c & c - 46) >>> 8 & -INVALID_BYTE + c - 45 + 62;
          result += (94 - c & c - 96) >>> 8 & -INVALID_BYTE + c - 95 + 63;
          result += (47 - c & c - 58) >>> 8 & -INVALID_BYTE + c - 48 + 52;
          result += (64 - c & c - 91) >>> 8 & -INVALID_BYTE + c - 65 + 0;
          result += (96 - c & c - 123) >>> 8 & -INVALID_BYTE + c - 97 + 26;
          return result;
        };
        return URLSafeCoder2;
      }(Coder)
    );
    exports2.URLSafeCoder = URLSafeCoder;
    var urlSafeCoder = new URLSafeCoder();
    function encodeURLSafe(data) {
      return urlSafeCoder.encode(data);
    }
    exports2.encodeURLSafe = encodeURLSafe;
    function decodeURLSafe(s) {
      return urlSafeCoder.decode(s);
    }
    exports2.decodeURLSafe = decodeURLSafe;
    exports2.encodedLength = function(length) {
      return stdCoder.encodedLength(length);
    };
    exports2.maxDecodedLength = function(length) {
      return stdCoder.maxDecodedLength(length);
    };
    exports2.decodedLength = function(s) {
      return stdCoder.decodedLength(s);
    };
  }
});

// ../../node_modules/fast-sha256/sha256.js
var require_sha256 = __commonJS({
  "../../node_modules/fast-sha256/sha256.js"(exports2, module2) {
    "use strict";
    (function(root, factory) {
      var exports3 = {};
      factory(exports3);
      var sha2562 = exports3["default"];
      for (var k in exports3) {
        sha2562[k] = exports3[k];
      }
      if (typeof module2 === "object" && typeof module2.exports === "object") {
        module2.exports = sha2562;
      } else if (typeof define === "function" && define.amd) {
        define(function() {
          return sha2562;
        });
      } else {
        root.sha256 = sha2562;
      }
    })(exports2, function(exports3) {
      "use strict";
      exports3.__esModule = true;
      exports3.digestLength = 32;
      exports3.blockSize = 64;
      var K = new Uint32Array([
        1116352408,
        1899447441,
        3049323471,
        3921009573,
        961987163,
        1508970993,
        2453635748,
        2870763221,
        3624381080,
        310598401,
        607225278,
        1426881987,
        1925078388,
        2162078206,
        2614888103,
        3248222580,
        3835390401,
        4022224774,
        264347078,
        604807628,
        770255983,
        1249150122,
        1555081692,
        1996064986,
        2554220882,
        2821834349,
        2952996808,
        3210313671,
        3336571891,
        3584528711,
        113926993,
        338241895,
        666307205,
        773529912,
        1294757372,
        1396182291,
        1695183700,
        1986661051,
        2177026350,
        2456956037,
        2730485921,
        2820302411,
        3259730800,
        3345764771,
        3516065817,
        3600352804,
        4094571909,
        275423344,
        430227734,
        506948616,
        659060556,
        883997877,
        958139571,
        1322822218,
        1537002063,
        1747873779,
        1955562222,
        2024104815,
        2227730452,
        2361852424,
        2428436474,
        2756734187,
        3204031479,
        3329325298
      ]);
      function hashBlocks(w, v, p, pos, len) {
        var a, b, c, d, e, f, g, h, u, i, j, t1, t2;
        while (len >= 64) {
          a = v[0];
          b = v[1];
          c = v[2];
          d = v[3];
          e = v[4];
          f = v[5];
          g = v[6];
          h = v[7];
          for (i = 0; i < 16; i++) {
            j = pos + i * 4;
            w[i] = (p[j] & 255) << 24 | (p[j + 1] & 255) << 16 | (p[j + 2] & 255) << 8 | p[j + 3] & 255;
          }
          for (i = 16; i < 64; i++) {
            u = w[i - 2];
            t1 = (u >>> 17 | u << 32 - 17) ^ (u >>> 19 | u << 32 - 19) ^ u >>> 10;
            u = w[i - 15];
            t2 = (u >>> 7 | u << 32 - 7) ^ (u >>> 18 | u << 32 - 18) ^ u >>> 3;
            w[i] = (t1 + w[i - 7] | 0) + (t2 + w[i - 16] | 0);
          }
          for (i = 0; i < 64; i++) {
            t1 = (((e >>> 6 | e << 32 - 6) ^ (e >>> 11 | e << 32 - 11) ^ (e >>> 25 | e << 32 - 25)) + (e & f ^ ~e & g) | 0) + (h + (K[i] + w[i] | 0) | 0) | 0;
            t2 = ((a >>> 2 | a << 32 - 2) ^ (a >>> 13 | a << 32 - 13) ^ (a >>> 22 | a << 32 - 22)) + (a & b ^ a & c ^ b & c) | 0;
            h = g;
            g = f;
            f = e;
            e = d + t1 | 0;
            d = c;
            c = b;
            b = a;
            a = t1 + t2 | 0;
          }
          v[0] += a;
          v[1] += b;
          v[2] += c;
          v[3] += d;
          v[4] += e;
          v[5] += f;
          v[6] += g;
          v[7] += h;
          pos += 64;
          len -= 64;
        }
        return pos;
      }
      var Hash = (
        /** @class */
        function() {
          function Hash2() {
            this.digestLength = exports3.digestLength;
            this.blockSize = exports3.blockSize;
            this.state = new Int32Array(8);
            this.temp = new Int32Array(64);
            this.buffer = new Uint8Array(128);
            this.bufferLength = 0;
            this.bytesHashed = 0;
            this.finished = false;
            this.reset();
          }
          Hash2.prototype.reset = function() {
            this.state[0] = 1779033703;
            this.state[1] = 3144134277;
            this.state[2] = 1013904242;
            this.state[3] = 2773480762;
            this.state[4] = 1359893119;
            this.state[5] = 2600822924;
            this.state[6] = 528734635;
            this.state[7] = 1541459225;
            this.bufferLength = 0;
            this.bytesHashed = 0;
            this.finished = false;
            return this;
          };
          Hash2.prototype.clean = function() {
            for (var i = 0; i < this.buffer.length; i++) {
              this.buffer[i] = 0;
            }
            for (var i = 0; i < this.temp.length; i++) {
              this.temp[i] = 0;
            }
            this.reset();
          };
          Hash2.prototype.update = function(data, dataLength) {
            if (dataLength === void 0) {
              dataLength = data.length;
            }
            if (this.finished) {
              throw new Error("SHA256: can't update because hash was finished.");
            }
            var dataPos = 0;
            this.bytesHashed += dataLength;
            if (this.bufferLength > 0) {
              while (this.bufferLength < 64 && dataLength > 0) {
                this.buffer[this.bufferLength++] = data[dataPos++];
                dataLength--;
              }
              if (this.bufferLength === 64) {
                hashBlocks(this.temp, this.state, this.buffer, 0, 64);
                this.bufferLength = 0;
              }
            }
            if (dataLength >= 64) {
              dataPos = hashBlocks(this.temp, this.state, data, dataPos, dataLength);
              dataLength %= 64;
            }
            while (dataLength > 0) {
              this.buffer[this.bufferLength++] = data[dataPos++];
              dataLength--;
            }
            return this;
          };
          Hash2.prototype.finish = function(out) {
            if (!this.finished) {
              var bytesHashed = this.bytesHashed;
              var left = this.bufferLength;
              var bitLenHi = bytesHashed / 536870912 | 0;
              var bitLenLo = bytesHashed << 3;
              var padLength = bytesHashed % 64 < 56 ? 64 : 128;
              this.buffer[left] = 128;
              for (var i = left + 1; i < padLength - 8; i++) {
                this.buffer[i] = 0;
              }
              this.buffer[padLength - 8] = bitLenHi >>> 24 & 255;
              this.buffer[padLength - 7] = bitLenHi >>> 16 & 255;
              this.buffer[padLength - 6] = bitLenHi >>> 8 & 255;
              this.buffer[padLength - 5] = bitLenHi >>> 0 & 255;
              this.buffer[padLength - 4] = bitLenLo >>> 24 & 255;
              this.buffer[padLength - 3] = bitLenLo >>> 16 & 255;
              this.buffer[padLength - 2] = bitLenLo >>> 8 & 255;
              this.buffer[padLength - 1] = bitLenLo >>> 0 & 255;
              hashBlocks(this.temp, this.state, this.buffer, 0, padLength);
              this.finished = true;
            }
            for (var i = 0; i < 8; i++) {
              out[i * 4 + 0] = this.state[i] >>> 24 & 255;
              out[i * 4 + 1] = this.state[i] >>> 16 & 255;
              out[i * 4 + 2] = this.state[i] >>> 8 & 255;
              out[i * 4 + 3] = this.state[i] >>> 0 & 255;
            }
            return this;
          };
          Hash2.prototype.digest = function() {
            var out = new Uint8Array(this.digestLength);
            this.finish(out);
            return out;
          };
          Hash2.prototype._saveState = function(out) {
            for (var i = 0; i < this.state.length; i++) {
              out[i] = this.state[i];
            }
          };
          Hash2.prototype._restoreState = function(from, bytesHashed) {
            for (var i = 0; i < this.state.length; i++) {
              this.state[i] = from[i];
            }
            this.bytesHashed = bytesHashed;
            this.finished = false;
            this.bufferLength = 0;
          };
          return Hash2;
        }()
      );
      exports3.Hash = Hash;
      var HMAC = (
        /** @class */
        function() {
          function HMAC2(key) {
            this.inner = new Hash();
            this.outer = new Hash();
            this.blockSize = this.inner.blockSize;
            this.digestLength = this.inner.digestLength;
            var pad = new Uint8Array(this.blockSize);
            if (key.length > this.blockSize) {
              new Hash().update(key).finish(pad).clean();
            } else {
              for (var i = 0; i < key.length; i++) {
                pad[i] = key[i];
              }
            }
            for (var i = 0; i < pad.length; i++) {
              pad[i] ^= 54;
            }
            this.inner.update(pad);
            for (var i = 0; i < pad.length; i++) {
              pad[i] ^= 54 ^ 92;
            }
            this.outer.update(pad);
            this.istate = new Uint32Array(8);
            this.ostate = new Uint32Array(8);
            this.inner._saveState(this.istate);
            this.outer._saveState(this.ostate);
            for (var i = 0; i < pad.length; i++) {
              pad[i] = 0;
            }
          }
          HMAC2.prototype.reset = function() {
            this.inner._restoreState(this.istate, this.inner.blockSize);
            this.outer._restoreState(this.ostate, this.outer.blockSize);
            return this;
          };
          HMAC2.prototype.clean = function() {
            for (var i = 0; i < this.istate.length; i++) {
              this.ostate[i] = this.istate[i] = 0;
            }
            this.inner.clean();
            this.outer.clean();
          };
          HMAC2.prototype.update = function(data) {
            this.inner.update(data);
            return this;
          };
          HMAC2.prototype.finish = function(out) {
            if (this.outer.finished) {
              this.outer.finish(out);
            } else {
              this.inner.finish(out);
              this.outer.update(out, this.digestLength).finish(out);
            }
            return this;
          };
          HMAC2.prototype.digest = function() {
            var out = new Uint8Array(this.digestLength);
            this.finish(out);
            return out;
          };
          return HMAC2;
        }()
      );
      exports3.HMAC = HMAC;
      function hash(data) {
        var h = new Hash().update(data);
        var digest = h.digest();
        h.clean();
        return digest;
      }
      exports3.hash = hash;
      exports3["default"] = hash;
      function hmac2(key, data) {
        var h = new HMAC(key).update(data);
        var digest = h.digest();
        h.clean();
        return digest;
      }
      exports3.hmac = hmac2;
      function fillBuffer(buffer, hmac3, info, counter) {
        var num = counter[0];
        if (num === 0) {
          throw new Error("hkdf: cannot expand more");
        }
        hmac3.reset();
        if (num > 1) {
          hmac3.update(buffer);
        }
        if (info) {
          hmac3.update(info);
        }
        hmac3.update(counter);
        hmac3.finish(buffer);
        counter[0]++;
      }
      var hkdfSalt = new Uint8Array(exports3.digestLength);
      function hkdf(key, salt, info, length) {
        if (salt === void 0) {
          salt = hkdfSalt;
        }
        if (length === void 0) {
          length = 32;
        }
        var counter = new Uint8Array([1]);
        var okm = hmac2(salt, key);
        var hmac_ = new HMAC(okm);
        var buffer = new Uint8Array(hmac_.digestLength);
        var bufpos = buffer.length;
        var out = new Uint8Array(length);
        for (var i = 0; i < length; i++) {
          if (bufpos === buffer.length) {
            fillBuffer(buffer, hmac_, info, counter);
            bufpos = 0;
          }
          out[i] = buffer[bufpos++];
        }
        hmac_.clean();
        buffer.fill(0);
        counter.fill(0);
        return out;
      }
      exports3.hkdf = hkdf;
      function pbkdf2(password, salt, iterations, dkLen) {
        var prf = new HMAC(password);
        var len = prf.digestLength;
        var ctr = new Uint8Array(4);
        var t = new Uint8Array(len);
        var u = new Uint8Array(len);
        var dk = new Uint8Array(dkLen);
        for (var i = 0; i * len < dkLen; i++) {
          var c = i + 1;
          ctr[0] = c >>> 24 & 255;
          ctr[1] = c >>> 16 & 255;
          ctr[2] = c >>> 8 & 255;
          ctr[3] = c >>> 0 & 255;
          prf.reset();
          prf.update(salt);
          prf.update(ctr);
          prf.finish(u);
          for (var j = 0; j < len; j++) {
            t[j] = u[j];
          }
          for (var j = 2; j <= iterations; j++) {
            prf.reset();
            prf.update(u).finish(u);
            for (var k = 0; k < len; k++) {
              t[k] ^= u[k];
            }
          }
          for (var j = 0; j < len && i * len + j < dkLen; j++) {
            dk[i * len + j] = t[j];
          }
        }
        for (var i = 0; i < len; i++) {
          t[i] = u[i] = 0;
        }
        for (var i = 0; i < 4; i++) {
          ctr[i] = 0;
        }
        prf.clean();
        return dk;
      }
      exports3.pbkdf2 = pbkdf2;
    });
  }
});

// src/webhook/webhook.ts
var webhook_exports = {};
__export(webhook_exports, {
  handleWebhookPayload: () => handleWebhookPayload,
  verifyWebhookPayload: () => verifyWebhookPayload
});
module.exports = __toCommonJS(webhook_exports);

// src/schemas/webhook.ts
var import_v3 = require("zod/v3");
var PaymentSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("Payment"),
  billing: import_v3.z.object({
    city: import_v3.z.string().nullable(),
    country: import_v3.z.string(),
    state: import_v3.z.string().nullable(),
    street: import_v3.z.string().nullable(),
    zipcode: import_v3.z.string().nullable()
  }),
  brand_id: import_v3.z.string(),
  business_id: import_v3.z.string(),
  card_holder_name: import_v3.z.string().nullable(),
  card_issuing_country: import_v3.z.string().nullable(),
  card_last_four: import_v3.z.string().nullable(),
  card_network: import_v3.z.string().nullable(),
  card_type: import_v3.z.string().nullable(),
  checkout_session_id: import_v3.z.string().nullable(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  currency: import_v3.z.string(),
  custom_field_responses: import_v3.z.array(
    import_v3.z.object({
      key: import_v3.z.string(),
      value: import_v3.z.string()
    })
  ).nullable(),
  customer: import_v3.z.object({
    customer_id: import_v3.z.string(),
    email: import_v3.z.string(),
    metadata: import_v3.z.record(import_v3.z.any()),
    name: import_v3.z.string(),
    phone_number: import_v3.z.string().nullable()
  }),
  digital_products_delivered: import_v3.z.boolean(),
  /**
   * @deprecated Use `discount_ids` instead. Still populated for backward
   * compatibility with the singular discount_code flow.
   */
  discount_id: import_v3.z.string().nullable(),
  /**
   * Flat list of all stacked discount IDs applied to the payment, in order
   * of application. Always present on v1.98.0+ webhook payloads.
   */
  discount_ids: import_v3.z.array(import_v3.z.string()).nullable().optional(),
  disputes: import_v3.z.array(
    import_v3.z.object({
      amount: import_v3.z.string(),
      business_id: import_v3.z.string(),
      created_at: import_v3.z.string().transform((d) => new Date(d)),
      currency: import_v3.z.string(),
      dispute_id: import_v3.z.string(),
      dispute_stage: import_v3.z.enum(["pre_dispute", "dispute", "pre_arbitration"]),
      dispute_status: import_v3.z.enum([
        "dispute_opened",
        "dispute_expired",
        "dispute_accepted",
        "dispute_cancelled",
        "dispute_challenged",
        "dispute_won",
        "dispute_lost"
      ]),
      payment_id: import_v3.z.string(),
      remarks: import_v3.z.string().nullable()
    })
  ).default([]),
  error_code: import_v3.z.string().nullable(),
  error_message: import_v3.z.string().nullable(),
  invoice_id: import_v3.z.string().nullable(),
  invoice_url: import_v3.z.string().nullable(),
  metadata: import_v3.z.record(import_v3.z.any()),
  payment_id: import_v3.z.string(),
  payment_link: import_v3.z.string().nullable(),
  payment_method: import_v3.z.string().nullable(),
  payment_method_type: import_v3.z.string().nullable(),
  product_cart: import_v3.z.array(
    import_v3.z.object({
      product_id: import_v3.z.string(),
      quantity: import_v3.z.number()
    })
  ).nullable(),
  refunds: import_v3.z.array(
    import_v3.z.object({
      amount: import_v3.z.number().nullable(),
      business_id: import_v3.z.string(),
      created_at: import_v3.z.string().transform((d) => new Date(d)),
      currency: import_v3.z.string().nullable(),
      is_partial: import_v3.z.boolean(),
      payment_id: import_v3.z.string(),
      reason: import_v3.z.string().nullable(),
      refund_id: import_v3.z.string(),
      status: import_v3.z.enum(["succeeded", "failed", "pending", "review"])
    })
  ),
  refund_status: import_v3.z.enum(["partial", "full"]).nullable(),
  settlement_amount: import_v3.z.number(),
  settlement_currency: import_v3.z.string(),
  settlement_tax: import_v3.z.number().nullable(),
  status: import_v3.z.enum([
    "succeeded",
    "failed",
    "cancelled",
    "processing",
    "requires_customer_action",
    "requires_merchant_action",
    "requires_payment_method",
    "requires_confirmation",
    "requires_capture",
    "partially_captured",
    "partially_captured_and_capturable"
  ]).nullable(),
  subscription_id: import_v3.z.string().nullable(),
  tax: import_v3.z.number().nullable(),
  total_amount: import_v3.z.number(),
  updated_at: import_v3.z.string().transform((d) => new Date(d)).nullable()
});
var SubscriptionSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("Subscription"),
  addons: import_v3.z.array(
    import_v3.z.object({
      addon_id: import_v3.z.string(),
      quantity: import_v3.z.number()
    })
  ),
  billing: import_v3.z.object({
    city: import_v3.z.string().nullable(),
    country: import_v3.z.string(),
    state: import_v3.z.string().nullable(),
    street: import_v3.z.string().nullable(),
    zipcode: import_v3.z.string().nullable()
  }),
  cancel_at_next_billing_date: import_v3.z.boolean(),
  cancelled_at: import_v3.z.string().transform((d) => new Date(d)).nullable(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  currency: import_v3.z.string(),
  customer: import_v3.z.object({
    customer_id: import_v3.z.string(),
    email: import_v3.z.string(),
    metadata: import_v3.z.record(import_v3.z.any()),
    name: import_v3.z.string(),
    phone_number: import_v3.z.string().nullable()
  }),
  custom_field_responses: import_v3.z.array(
    import_v3.z.object({
      key: import_v3.z.string(),
      value: import_v3.z.string()
    })
  ).nullable(),
  /**
   * @deprecated Use `discounts[].discount_cycles_remaining` instead. Still
   * populated for backward compatibility with the singular discount_code
   * flow.
   */
  discount_cycles_remaining: import_v3.z.number().nullable(),
  /**
   * @deprecated Use `discounts` (full objects) or `discount_ids` (flat list)
   * instead. Still populated for backward compatibility with the singular
   * discount_code flow.
   */
  discount_id: import_v3.z.string().nullable(),
  /**
   * Flat list of all stacked discount IDs applied to the subscription, in
   * order of application. Always present on v1.98.0+ webhook payloads.
   */
  discount_ids: import_v3.z.array(import_v3.z.string()).nullable().optional(),
  /**
   * All stacked discounts applied to the subscription, in order of
   * application. Each entry contains the discount's position, remaining
   * billing cycles, code, and metadata. Always present on v1.98.0+ webhook
   * payloads.
   */
  discounts: import_v3.z.array(
    import_v3.z.object({
      discount_id: import_v3.z.string(),
      code: import_v3.z.string().nullable().optional(),
      position: import_v3.z.number().int().nonnegative(),
      cycles_remaining: import_v3.z.number().nullable().optional(),
      metadata: import_v3.z.record(import_v3.z.any()).nullable().optional()
    })
  ).nullable().optional(),
  expires_at: import_v3.z.string().transform((d) => new Date(d)).nullable(),
  credit_entitlement_cart: import_v3.z.array(
    import_v3.z.object({
      credit_entitlement_id: import_v3.z.string(),
      credit_entitlement_name: import_v3.z.string(),
      credits_amount: import_v3.z.string(),
      overage_balance: import_v3.z.string(),
      overage_behavior: import_v3.z.enum([
        "forgive_at_reset",
        "invoice_at_billing",
        "carry_deficit",
        "carry_deficit_auto_repay"
      ]),
      overage_enabled: import_v3.z.boolean(),
      product_id: import_v3.z.string(),
      remaining_balance: import_v3.z.string(),
      rollover_enabled: import_v3.z.boolean(),
      unit: import_v3.z.string(),
      expires_after_days: import_v3.z.number().nullable(),
      low_balance_threshold_percent: import_v3.z.number().nullable(),
      max_rollover_count: import_v3.z.number().nullable(),
      overage_limit: import_v3.z.string().nullable(),
      rollover_percentage: import_v3.z.number().nullable(),
      rollover_timeframe_count: import_v3.z.number().nullable(),
      rollover_timeframe_interval: import_v3.z.enum(["Day", "Week", "Month", "Year"]).nullable()
    })
  ),
  meter_credit_entitlement_cart: import_v3.z.array(
    import_v3.z.object({
      credit_entitlement_id: import_v3.z.string(),
      meter_id: import_v3.z.string(),
      meter_name: import_v3.z.string(),
      meter_units_per_credit: import_v3.z.string(),
      product_id: import_v3.z.string()
    })
  ),
  meters: import_v3.z.array(
    import_v3.z.object({
      currency: import_v3.z.string(),
      description: import_v3.z.string().nullable(),
      free_threshold: import_v3.z.number(),
      measurement_unit: import_v3.z.string(),
      meter_id: import_v3.z.string(),
      name: import_v3.z.string(),
      price_per_unit: import_v3.z.string().nullable()
    })
  ),
  metadata: import_v3.z.record(import_v3.z.any()),
  next_billing_date: import_v3.z.string().transform((d) => new Date(d)),
  on_demand: import_v3.z.boolean(),
  payment_frequency_count: import_v3.z.number(),
  payment_frequency_interval: import_v3.z.enum(["Day", "Week", "Month", "Year"]),
  payment_method_id: import_v3.z.string().nullable(),
  previous_billing_date: import_v3.z.string().transform((d) => new Date(d)),
  product_id: import_v3.z.string(),
  quantity: import_v3.z.number(),
  recurring_pre_tax_amount: import_v3.z.number(),
  status: import_v3.z.enum([
    "pending",
    "active",
    "on_hold",
    "cancelled",
    "expired",
    "failed"
  ]),
  subscription_id: import_v3.z.string(),
  subscription_period_count: import_v3.z.number(),
  subscription_period_interval: import_v3.z.enum(["Day", "Week", "Month", "Year"]),
  tax_id: import_v3.z.string().nullable(),
  tax_inclusive: import_v3.z.boolean(),
  trial_period_days: import_v3.z.number()
});
var RefundSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("Refund"),
  amount: import_v3.z.number().nullable(),
  business_id: import_v3.z.string(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  customer: import_v3.z.object({
    customer_id: import_v3.z.string(),
    email: import_v3.z.string(),
    metadata: import_v3.z.record(import_v3.z.any()),
    name: import_v3.z.string(),
    phone_number: import_v3.z.string().nullable()
  }),
  currency: import_v3.z.string().nullable(),
  is_partial: import_v3.z.boolean(),
  metadata: import_v3.z.record(import_v3.z.any()),
  payment_id: import_v3.z.string(),
  reason: import_v3.z.string().nullable(),
  refund_id: import_v3.z.string(),
  status: import_v3.z.enum(["succeeded", "failed", "pending", "review"])
});
var DisputeSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("Dispute"),
  amount: import_v3.z.string(),
  business_id: import_v3.z.string(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  currency: import_v3.z.string(),
  customer: import_v3.z.object({
    customer_id: import_v3.z.string(),
    email: import_v3.z.string(),
    metadata: import_v3.z.record(import_v3.z.any()),
    name: import_v3.z.string(),
    phone_number: import_v3.z.string().nullable()
  }),
  dispute_id: import_v3.z.string(),
  dispute_stage: import_v3.z.enum(["pre_dispute", "dispute", "pre_arbitration"]),
  dispute_status: import_v3.z.enum([
    "dispute_opened",
    "dispute_expired",
    "dispute_accepted",
    "dispute_cancelled",
    "dispute_challenged",
    "dispute_won",
    "dispute_lost"
  ]),
  payment_id: import_v3.z.string(),
  reason: import_v3.z.string().nullable(),
  remarks: import_v3.z.string().nullable()
});
var LicenseKeySchema = import_v3.z.object({
  payload_type: import_v3.z.literal("LicenseKey"),
  activations_limit: import_v3.z.number().nullable(),
  business_id: import_v3.z.string(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  customer_id: import_v3.z.string(),
  expires_at: import_v3.z.string().transform((d) => new Date(d)).nullable(),
  id: import_v3.z.string(),
  instances_count: import_v3.z.number(),
  key: import_v3.z.string(),
  payment_id: import_v3.z.string(),
  product_id: import_v3.z.string(),
  status: import_v3.z.enum(["active", "expired", "disabled"]),
  subscription_id: import_v3.z.string().nullable()
});
var PaymentSucceededPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("payment.succeeded"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: PaymentSchema
});
var PaymentFailedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("payment.failed"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: PaymentSchema
});
var PaymentProcessingPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("payment.processing"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: PaymentSchema
});
var PaymentCancelledPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("payment.cancelled"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: PaymentSchema
});
var RefundSucceededPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("refund.succeeded"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: RefundSchema
});
var RefundFailedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("refund.failed"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: RefundSchema
});
var DisputeOpenedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.opened"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeExpiredPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.expired"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeAcceptedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.accepted"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeCancelledPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.cancelled"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeChallengedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.challenged"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeWonPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.won"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeLostPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.lost"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var SubscriptionActivePayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.active"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionOnHoldPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.on_hold"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionRenewedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.renewed"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionPlanChangedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.plan_changed"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionCancelledPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.cancelled"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionCancellationScheduledPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.cancellation_scheduled"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionTrialEndingPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.trial_ending"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionUpcomingRenewalPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.upcoming_renewal"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionFailedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.failed"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionExpiredPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.expired"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionUpdatedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.updated"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var LicenseKeyCreatedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("license_key.created"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: LicenseKeySchema
});
var CreditLedgerEntrySchema = import_v3.z.object({
  payload_type: import_v3.z.literal("CreditLedgerEntry"),
  id: import_v3.z.string(),
  amount: import_v3.z.string(),
  balance_after: import_v3.z.string(),
  balance_before: import_v3.z.string(),
  business_id: import_v3.z.string(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  credit_entitlement_id: import_v3.z.string(),
  customer_id: import_v3.z.string(),
  is_credit: import_v3.z.boolean(),
  overage_after: import_v3.z.string(),
  overage_before: import_v3.z.string(),
  transaction_type: import_v3.z.enum([
    "credit_added",
    "credit_deducted",
    "credit_expired",
    "credit_rolled_over",
    "rollover_forfeited",
    "overage_charged",
    "auto_top_up",
    "manual_adjustment",
    "refund"
  ]),
  description: import_v3.z.string().nullable(),
  grant_id: import_v3.z.string().nullable(),
  reference_id: import_v3.z.string().nullable(),
  reference_type: import_v3.z.string().nullable()
});
var CreditBalanceLowSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("CreditBalanceLow"),
  customer_id: import_v3.z.string(),
  subscription_id: import_v3.z.string(),
  credit_entitlement_id: import_v3.z.string(),
  credit_entitlement_name: import_v3.z.string(),
  available_balance: import_v3.z.string(),
  subscription_credits_amount: import_v3.z.string(),
  threshold_percent: import_v3.z.number(),
  threshold_amount: import_v3.z.string()
});
var CreditAddedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.added"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditDeductedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.deducted"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditExpiredPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.expired"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditRolledOverPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.rolled_over"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditRolloverForfeitedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.rollover_forfeited"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditOverageChargedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.overage_charged"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditManualAdjustmentPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.manual_adjustment"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditBalanceLowPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.balance_low"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditBalanceLowSchema
});
var AbandonedCheckoutSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("AbandonedCheckout"),
  abandoned_at: import_v3.z.string().transform((d) => new Date(d)),
  abandonment_reason: import_v3.z.enum(["payment_failed", "checkout_incomplete"]),
  customer_id: import_v3.z.string(),
  payment_id: import_v3.z.string(),
  status: import_v3.z.enum(["abandoned", "recovering", "recovered", "exhausted", "opted_out"]),
  recovered_payment_id: import_v3.z.string().nullable().optional()
});
var DunningAttemptSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("DunningAttempt"),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  customer_id: import_v3.z.string(),
  status: import_v3.z.enum(["recovering", "recovered", "exhausted"]),
  subscription_id: import_v3.z.string(),
  trigger_state: import_v3.z.enum(["on_hold", "cancelled"]),
  payment_id: import_v3.z.string().nullable().optional()
});
var AbandonedCheckoutDetectedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("abandoned_checkout.detected"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: AbandonedCheckoutSchema
});
var AbandonedCheckoutRecoveredPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("abandoned_checkout.recovered"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: AbandonedCheckoutSchema
});
var DunningStartedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dunning.started"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DunningAttemptSchema
});
var DunningRecoveredPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dunning.recovered"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DunningAttemptSchema
});
var WebhookPayloadSchema = import_v3.z.discriminatedUnion("type", [
  PaymentSucceededPayloadSchema,
  PaymentFailedPayloadSchema,
  PaymentProcessingPayloadSchema,
  PaymentCancelledPayloadSchema,
  RefundSucceededPayloadSchema,
  RefundFailedPayloadSchema,
  DisputeOpenedPayloadSchema,
  DisputeExpiredPayloadSchema,
  DisputeAcceptedPayloadSchema,
  DisputeCancelledPayloadSchema,
  DisputeChallengedPayloadSchema,
  DisputeWonPayloadSchema,
  DisputeLostPayloadSchema,
  SubscriptionActivePayloadSchema,
  SubscriptionOnHoldPayloadSchema,
  SubscriptionRenewedPayloadSchema,
  SubscriptionPlanChangedPayloadSchema,
  SubscriptionCancelledPayloadSchema,
  SubscriptionCancellationScheduledPayloadSchema,
  SubscriptionTrialEndingPayloadSchema,
  SubscriptionUpcomingRenewalPayloadSchema,
  SubscriptionFailedPayloadSchema,
  SubscriptionExpiredPayloadSchema,
  SubscriptionUpdatedPayloadSchema,
  LicenseKeyCreatedPayloadSchema,
  AbandonedCheckoutDetectedPayloadSchema,
  AbandonedCheckoutRecoveredPayloadSchema,
  DunningStartedPayloadSchema,
  DunningRecoveredPayloadSchema,
  CreditAddedPayloadSchema,
  CreditDeductedPayloadSchema,
  CreditExpiredPayloadSchema,
  CreditRolledOverPayloadSchema,
  CreditRolloverForfeitedPayloadSchema,
  CreditOverageChargedPayloadSchema,
  CreditManualAdjustmentPayloadSchema,
  CreditBalanceLowPayloadSchema
]);

// src/webhook/vendor/timing_safe_equal.ts
function assert(expr, msg = "") {
  if (!expr) {
    throw new Error(msg);
  }
}
function timingSafeEqual(a, b) {
  if (a.byteLength !== b.byteLength) {
    return false;
  }
  if (!(a instanceof DataView)) {
    a = new DataView(ArrayBuffer.isView(a) ? a.buffer : a);
  }
  if (!(b instanceof DataView)) {
    b = new DataView(ArrayBuffer.isView(b) ? b.buffer : b);
  }
  assert(a instanceof DataView);
  assert(b instanceof DataView);
  const length = a.byteLength;
  let out = 0;
  let i = -1;
  while (++i < length) {
    out |= a.getUint8(i) ^ b.getUint8(i);
  }
  return out === 0;
}

// src/webhook/vendor/standardwebhook.ts
var base64 = __toESM(require_base64(), 1);
var sha256 = __toESM(require_sha256(), 1);
var WEBHOOK_TOLERANCE_IN_SECONDS = 5 * 60;
var ExtendableError = class _ExtendableError extends Error {
  constructor(message) {
    super(message);
    Object.setPrototypeOf(this, _ExtendableError.prototype);
    this.name = "ExtendableError";
    this.stack = new Error(message).stack;
  }
};
var WebhookVerificationError = class _WebhookVerificationError extends ExtendableError {
  constructor(message) {
    super(message);
    Object.setPrototypeOf(this, _WebhookVerificationError.prototype);
    this.name = "WebhookVerificationError";
  }
};
var Webhook = class _Webhook {
  static prefix = "whsec_";
  key;
  constructor(secret, options) {
    if (!secret) {
      throw new Error("Secret can't be empty.");
    }
    if ((options == null ? void 0 : options.format) === "raw") {
      if (secret instanceof Uint8Array) {
        this.key = secret;
      } else {
        this.key = Uint8Array.from(secret, (c) => c.charCodeAt(0));
      }
    } else {
      if (typeof secret !== "string") {
        throw new Error("Expected secret to be of type string");
      }
      if (secret.startsWith(_Webhook.prefix)) {
        secret = secret.substring(_Webhook.prefix.length);
      }
      this.key = base64.decode(secret);
    }
  }
  verify(payload, headers_) {
    const headers = {};
    for (const key of Object.keys(headers_)) {
      headers[key.toLowerCase()] = headers_[key];
    }
    const msgId = headers["webhook-id"];
    const msgSignature = headers["webhook-signature"];
    const msgTimestamp = headers["webhook-timestamp"];
    if (!msgSignature || !msgId || !msgTimestamp) {
      throw new WebhookVerificationError("Missing required headers");
    }
    const timestamp = this.verifyTimestamp(msgTimestamp);
    const computedSignature = this.sign(msgId, timestamp, payload);
    const expectedSignature = computedSignature.split(",")[1];
    const passedSignatures = msgSignature.split(" ");
    const encoder = new globalThis.TextEncoder();
    for (const versionedSignature of passedSignatures) {
      const [version, signature] = versionedSignature.split(",");
      if (version !== "v1") {
        continue;
      }
      if (timingSafeEqual(
        encoder.encode(signature),
        encoder.encode(expectedSignature)
      )) {
        return JSON.parse(payload.toString());
      }
    }
    throw new WebhookVerificationError("No matching signature found");
  }
  sign(msgId, timestamp, payload) {
    if (typeof payload === "string") {
    } else if (payload.constructor.name === "Buffer") {
      payload = payload.toString();
    } else {
      throw new Error("Expected payload to be of type string or Buffer.");
    }
    const encoder = new TextEncoder();
    const timestampNumber = Math.floor(timestamp.getTime() / 1e3);
    const toSign = encoder.encode(`${msgId}.${timestampNumber}.${payload}`);
    const expectedSignature = base64.encode(sha256.hmac(this.key, toSign));
    return `v1,${expectedSignature}`;
  }
  verifyTimestamp(timestampHeader) {
    const now = Math.floor(Date.now() / 1e3);
    const timestamp = parseInt(timestampHeader, 10);
    if (isNaN(timestamp)) {
      throw new WebhookVerificationError("Invalid Signature Headers");
    }
    if (now - timestamp > WEBHOOK_TOLERANCE_IN_SECONDS) {
      throw new WebhookVerificationError("Message timestamp too old");
    }
    if (timestamp > now + WEBHOOK_TOLERANCE_IN_SECONDS) {
      throw new WebhookVerificationError("Message timestamp too new");
    }
    return new Date(timestamp * 1e3);
  }
};

// src/webhook/webhook.ts
async function handleWebhookPayload(payload, config, context) {
  const callHandler = (handler, payload2) => {
    if (!handler) return;
    if (context !== void 0) {
      return handler(context, payload2);
    }
    return handler(payload2);
  };
  if (config.onPayload) {
    await callHandler(config.onPayload, payload);
  }
  if (payload.type === "payment.succeeded") {
    await callHandler(config.onPaymentSucceeded, payload);
  }
  if (payload.type === "payment.failed") {
    await callHandler(config.onPaymentFailed, payload);
  }
  if (payload.type === "payment.processing") {
    await callHandler(config.onPaymentProcessing, payload);
  }
  if (payload.type === "payment.cancelled") {
    await callHandler(config.onPaymentCancelled, payload);
  }
  if (payload.type === "refund.succeeded") {
    await callHandler(config.onRefundSucceeded, payload);
  }
  if (payload.type === "refund.failed") {
    await callHandler(config.onRefundFailed, payload);
  }
  if (payload.type === "dispute.opened") {
    await callHandler(config.onDisputeOpened, payload);
  }
  if (payload.type === "dispute.expired") {
    await callHandler(config.onDisputeExpired, payload);
  }
  if (payload.type === "dispute.accepted") {
    await callHandler(config.onDisputeAccepted, payload);
  }
  if (payload.type === "dispute.cancelled") {
    await callHandler(config.onDisputeCancelled, payload);
  }
  if (payload.type === "dispute.challenged") {
    await callHandler(config.onDisputeChallenged, payload);
  }
  if (payload.type === "dispute.won") {
    await callHandler(config.onDisputeWon, payload);
  }
  if (payload.type === "dispute.lost") {
    await callHandler(config.onDisputeLost, payload);
  }
  if (payload.type === "subscription.active") {
    await callHandler(config.onSubscriptionActive, payload);
  }
  if (payload.type === "subscription.on_hold") {
    await callHandler(config.onSubscriptionOnHold, payload);
  }
  if (payload.type === "subscription.renewed") {
    await callHandler(config.onSubscriptionRenewed, payload);
  }
  if (payload.type === "subscription.plan_changed") {
    await callHandler(config.onSubscriptionPlanChanged, payload);
  }
  if (payload.type === "subscription.cancelled") {
    await callHandler(config.onSubscriptionCancelled, payload);
  }
  if (payload.type === "subscription.cancellation_scheduled") {
    await callHandler(config.onSubscriptionCancellationScheduled, payload);
  }
  if (payload.type === "subscription.trial_ending") {
    await callHandler(config.onSubscriptionTrialEnding, payload);
  }
  if (payload.type === "subscription.upcoming_renewal") {
    await callHandler(config.onSubscriptionUpcomingRenewal, payload);
  }
  if (payload.type === "subscription.failed") {
    await callHandler(config.onSubscriptionFailed, payload);
  }
  if (payload.type === "subscription.expired") {
    await callHandler(config.onSubscriptionExpired, payload);
  }
  if (payload.type === "subscription.updated") {
    await callHandler(config.onSubscriptionUpdated, payload);
  }
  if (payload.type === "license_key.created") {
    await callHandler(config.onLicenseKeyCreated, payload);
  }
  if (payload.type === "abandoned_checkout.detected") {
    await callHandler(config.onAbandonedCheckoutDetected, payload);
  }
  if (payload.type === "abandoned_checkout.recovered") {
    await callHandler(config.onAbandonedCheckoutRecovered, payload);
  }
  if (payload.type === "dunning.started") {
    await callHandler(config.onDunningStarted, payload);
  }
  if (payload.type === "dunning.recovered") {
    await callHandler(config.onDunningRecovered, payload);
  }
  if (payload.type === "credit.added") {
    await callHandler(config.onCreditAdded, payload);
  }
  if (payload.type === "credit.deducted") {
    await callHandler(config.onCreditDeducted, payload);
  }
  if (payload.type === "credit.expired") {
    await callHandler(config.onCreditExpired, payload);
  }
  if (payload.type === "credit.rolled_over") {
    await callHandler(config.onCreditRolledOver, payload);
  }
  if (payload.type === "credit.rollover_forfeited") {
    await callHandler(config.onCreditRolloverForfeited, payload);
  }
  if (payload.type === "credit.overage_charged") {
    await callHandler(config.onCreditOverageCharged, payload);
  }
  if (payload.type === "credit.manual_adjustment") {
    await callHandler(config.onCreditManualAdjustment, payload);
  }
  if (payload.type === "credit.balance_low") {
    await callHandler(config.onCreditBalanceLow, payload);
  }
}
var verifyWebhookPayload = async ({
  webhookKey,
  headers,
  body
}) => {
  const standardWebhook = new Webhook(webhookKey);
  try {
    standardWebhook.verify(body, headers);
  } catch (e) {
    if (e instanceof WebhookVerificationError) {
      throw new Error(e.message);
    }
    throw e;
  }
  const {
    success,
    data: payload,
    error
  } = WebhookPayloadSchema.safeParse(JSON.parse(body));
  if (!success) {
    throw new Error(error.message);
  }
  return payload;
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  handleWebhookPayload,
  verifyWebhookPayload
});
/**
 * The MIT License
 *
 * Copyright (c) 2023 Svix (https://www.svix.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @fileoverview Server-only webhook verification implementation.
 * @description Vendored from standardwebhooks package to avoid bundling issues.
 * Uses Node.js crypto module - DO NOT import in client/browser code.
 * @license MIT
 * @internal
 */
