// @dodopayments/core/dist/checkout/checkout.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/checkout/checkout.ts
var checkout_exports = {};
__export(checkout_exports, {
  MAX_STACKED_DISCOUNT_CODES: () => MAX_STACKED_DISCOUNT_CODES,
  assertDiscountFieldsExclusive: () => assertDiscountFieldsExclusive,
  buildCheckoutUrl: () => buildCheckoutUrl,
  checkoutQuerySchema: () => checkoutQuerySchema,
  checkoutSessionBillingAddressSchema: () => checkoutSessionBillingAddressSchema,
  checkoutSessionCreditEntitlementOverrideSchema: () => checkoutSessionCreditEntitlementOverrideSchema,
  checkoutSessionCustomerSchema: () => checkoutSessionCustomerSchema,
  checkoutSessionCustomizationSchema: () => checkoutSessionCustomizationSchema,
  checkoutSessionFeatureFlagsSchema: () => checkoutSessionFeatureFlagsSchema,
  checkoutSessionOnDemandSchema: () => checkoutSessionOnDemandSchema,
  checkoutSessionPayloadSchema: () => checkoutSessionPayloadSchema,
  checkoutSessionProductCartItemSchema: () => checkoutSessionProductCartItemSchema,
  checkoutSessionResponseSchema: () => checkoutSessionResponseSchema,
  checkoutSessionSubscriptionDataSchema: () => checkoutSessionSubscriptionDataSchema,
  createCheckoutSession: () => createCheckoutSession,
  discountCodesSchema: () => discountCodesSchema,
  dynamicCheckoutBodySchema: () => dynamicCheckoutBodySchema,
  paymentMethodTypeSchema: () => paymentMethodTypeSchema
});
module.exports = __toCommonJS(checkout_exports);
var import_v3 = require("zod/v3");
var import_dodopayments = __toESM(require("dodopayments"), 1);
var checkoutQuerySchema = import_v3.z.object({
  productId: import_v3.z.string(),
  quantity: import_v3.z.string().optional(),
  // Customer fields
  fullName: import_v3.z.string().optional(),
  firstName: import_v3.z.string().optional(),
  lastName: import_v3.z.string().optional(),
  email: import_v3.z.string().optional(),
  country: import_v3.z.string().optional(),
  addressLine: import_v3.z.string().optional(),
  city: import_v3.z.string().optional(),
  state: import_v3.z.string().optional(),
  zipCode: import_v3.z.string().optional(),
  // Disable flags
  disableFullName: import_v3.z.string().optional(),
  disableFirstName: import_v3.z.string().optional(),
  disableLastName: import_v3.z.string().optional(),
  disableEmail: import_v3.z.string().optional(),
  disableCountry: import_v3.z.string().optional(),
  disableAddressLine: import_v3.z.string().optional(),
  disableCity: import_v3.z.string().optional(),
  disableState: import_v3.z.string().optional(),
  disableZipCode: import_v3.z.string().optional(),
  // Advanced controls
  paymentCurrency: import_v3.z.string().optional(),
  showCurrencySelector: import_v3.z.string().optional(),
  paymentAmount: import_v3.z.string().optional(),
  showDiscounts: import_v3.z.string().optional()
  // Metadata (allow any key starting with metadata_)
  // We'll handle metadata separately in the handler
}).catchall(import_v3.z.unknown());
var dynamicCheckoutBodySchema = import_v3.z.object({
  // For subscription
  product_id: import_v3.z.string().optional(),
  quantity: import_v3.z.number().optional(),
  // For one-time payment
  product_cart: import_v3.z.array(
    import_v3.z.object({
      product_id: import_v3.z.string(),
      quantity: import_v3.z.number()
    })
  ).optional(),
  // Common fields
  billing: import_v3.z.object({
    city: import_v3.z.string(),
    country: import_v3.z.string(),
    state: import_v3.z.string(),
    street: import_v3.z.string(),
    zipcode: import_v3.z.string()
  }),
  customer: import_v3.z.object({
    customer_id: import_v3.z.string().optional(),
    email: import_v3.z.string().optional(),
    name: import_v3.z.string().optional()
  }),
  discount_id: import_v3.z.string().optional(),
  addons: import_v3.z.array(
    import_v3.z.object({
      addon_id: import_v3.z.string(),
      quantity: import_v3.z.number()
    })
  ).optional(),
  metadata: import_v3.z.record(import_v3.z.string(), import_v3.z.string()).optional(),
  currency: import_v3.z.string().optional(),
  // Discount codes (legacy + stacked).
  // `discount_code` is deprecated but still supported for backward
  // compatibility. `discount_codes` is the new stacked-discount field (max
  // 20, applied in order). The two cannot be combined in the same request.
  discount_code: import_v3.z.string().optional(),
  // Lazy-evaluated so we can share the schema definition below.
  discount_codes: import_v3.z.array(import_v3.z.string().min(1, "Discount code cannot be empty")).max(20, "At most 20 stacked discount codes are allowed").optional()
  // Allow any additional fields (for future compatibility)
}).catchall(import_v3.z.unknown());
var MAX_STACKED_DISCOUNT_CODES = 20;
var discountCodesSchema = import_v3.z.array(import_v3.z.string().min(1, "Discount code cannot be empty")).max(
  MAX_STACKED_DISCOUNT_CODES,
  `At most ${MAX_STACKED_DISCOUNT_CODES} stacked discount codes are allowed`
);
var checkoutSessionCreditEntitlementOverrideSchema = import_v3.z.object({
  credit_entitlement_id: import_v3.z.string().min(1, "credit_entitlement_id is required"),
  credits_amount: import_v3.z.string().min(1, "credits_amount is required (string for precision)")
});
var checkoutSessionProductCartItemSchema = import_v3.z.object({
  product_id: import_v3.z.string().min(1, "Product ID is required"),
  quantity: import_v3.z.number().int().positive("Quantity must be a positive integer"),
  addons: import_v3.z.array(
    import_v3.z.object({
      addon_id: import_v3.z.string(),
      quantity: import_v3.z.number().int().nonnegative()
    })
  ).optional(),
  amount: import_v3.z.number().int().nonnegative(
    "Amount must be a non-negative integer (for pay-what-you-want products)"
  ).optional(),
  // Per-checkout-session credit entitlement overrides. Each entry overrides
  // the `credits_amount` granted by the referenced credit entitlement when
  // this checkout session is fulfilled. The credit_entitlement_id must
  // already be attached to the product.
  credit_entitlements: import_v3.z.array(checkoutSessionCreditEntitlementOverrideSchema).optional()
});
var checkoutSessionCustomerSchema = import_v3.z.union([
  import_v3.z.object({
    email: import_v3.z.string().email(),
    name: import_v3.z.string().min(1).optional(),
    phone_number: import_v3.z.string().optional()
  }),
  import_v3.z.object({
    customer_id: import_v3.z.string()
  })
]).optional();
var checkoutSessionBillingAddressSchema = import_v3.z.object({
  street: import_v3.z.string().optional(),
  city: import_v3.z.string().optional(),
  state: import_v3.z.string().optional(),
  country: import_v3.z.string().length(2, "Country must be a 2-letter ISO code"),
  zipcode: import_v3.z.string().optional()
}).optional();
var paymentMethodTypeSchema = import_v3.z.enum([
  "credit",
  "debit",
  "upi_collect",
  "upi_intent",
  "apple_pay",
  "cashapp",
  "google_pay",
  "multibanco",
  "bancontact_card",
  "eps",
  "ideal",
  "przelewy24",
  "paypal",
  "affirm",
  "klarna",
  "sepa",
  "ach",
  "amazon_pay",
  "afterpay_clearpay"
]);
var checkoutSessionCustomizationSchema = import_v3.z.object({
  theme: import_v3.z.enum(["light", "dark", "system"]).optional(),
  show_order_details: import_v3.z.boolean().optional(),
  show_on_demand_tag: import_v3.z.boolean().optional(),
  force_language: import_v3.z.string().optional()
}).optional();
var checkoutSessionFeatureFlagsSchema = import_v3.z.object({
  allow_currency_selection: import_v3.z.boolean().optional(),
  allow_customer_editing_city: import_v3.z.boolean().optional(),
  allow_customer_editing_country: import_v3.z.boolean().optional(),
  allow_customer_editing_email: import_v3.z.boolean().optional(),
  allow_customer_editing_name: import_v3.z.boolean().optional(),
  allow_customer_editing_state: import_v3.z.boolean().optional(),
  allow_customer_editing_street: import_v3.z.boolean().optional(),
  allow_customer_editing_zipcode: import_v3.z.boolean().optional(),
  allow_discount_code: import_v3.z.boolean().optional(),
  allow_phone_number_collection: import_v3.z.boolean().optional(),
  allow_tax_id: import_v3.z.boolean().optional(),
  always_create_new_customer: import_v3.z.boolean().optional()
}).optional();
var checkoutSessionOnDemandSchema = import_v3.z.object({
  mandate_only: import_v3.z.boolean(),
  product_price: import_v3.z.number().int().optional(),
  product_currency: import_v3.z.string().length(3).optional(),
  product_description: import_v3.z.string().optional(),
  adaptive_currency_fees_inclusive: import_v3.z.boolean().optional()
}).optional();
var checkoutSessionSubscriptionDataSchema = import_v3.z.object({
  trial_period_days: import_v3.z.number().int().nonnegative().optional(),
  on_demand: checkoutSessionOnDemandSchema
}).optional();
var checkoutSessionPayloadSchema = import_v3.z.object({
  // Required fields
  product_cart: import_v3.z.array(checkoutSessionProductCartItemSchema).min(1, "At least one product is required"),
  // Optional fields
  customer: checkoutSessionCustomerSchema,
  billing_address: checkoutSessionBillingAddressSchema,
  return_url: import_v3.z.string().url().optional(),
  allowed_payment_method_types: import_v3.z.array(paymentMethodTypeSchema).optional(),
  billing_currency: import_v3.z.string().length(3, "Currency must be a 3-letter ISO code").optional(),
  show_saved_payment_methods: import_v3.z.boolean().optional(),
  confirm: import_v3.z.boolean().optional(),
  /**
   * @deprecated Use `discount_codes` instead. The singular `discount_code`
   * field continues to work for backward compatibility but cannot be
   * combined with the new `discount_codes` array in the same request.
   */
  discount_code: import_v3.z.string().optional(),
  /**
   * Stacked discount codes to apply, in order of application. Up to 20
   * codes. Cannot be combined with the deprecated singular `discount_code`
   * in the same request.
   */
  discount_codes: discountCodesSchema.optional(),
  metadata: import_v3.z.record(import_v3.z.string(), import_v3.z.string()).optional(),
  customization: checkoutSessionCustomizationSchema,
  feature_flags: checkoutSessionFeatureFlagsSchema,
  subscription_data: checkoutSessionSubscriptionDataSchema,
  force_3ds: import_v3.z.boolean().optional()
});
function assertDiscountFieldsExclusive(input) {
  if (input.discount_code != null && input.discount_code !== "" && input.discount_codes != null && input.discount_codes.length > 0) {
    throw new Error(
      "Cannot use both `discount_code` and `discount_codes` in the same request. The singular `discount_code` is deprecated \u2014 prefer `discount_codes`."
    );
  }
}
var checkoutSessionResponseSchema = import_v3.z.object({
  session_id: import_v3.z.string().min(1, "Session ID is required"),
  checkout_url: import_v3.z.string().url("Invalid checkout URL")
});
var createCheckoutSession = async (payload, config) => {
  const validation = checkoutSessionPayloadSchema.safeParse(payload);
  if (!validation.success) {
    throw new Error(
      `Invalid checkout session payload: ${validation.error.issues.map((issue) => `${issue.path.join(".")}: ${issue.message}`).join(", ")}`
    );
  }
  assertDiscountFieldsExclusive(validation.data);
  const dodopayments = new import_dodopayments.default({
    bearerToken: config.bearerToken,
    environment: config.environment
  });
  try {
    const sdkPayload = {
      ...validation.data,
      ...validation.data.billing_address && {
        billing_address: {
          ...validation.data.billing_address,
          country: validation.data.billing_address.country
        }
      }
    };
    const session = await dodopayments.checkoutSessions.create(
      sdkPayload
    );
    const responseValidation = checkoutSessionResponseSchema.safeParse(session);
    if (!responseValidation.success) {
      throw new Error(
        `Invalid checkout session response from API: ${responseValidation.error.issues.map((issue) => `${issue.path.join(".")}: ${issue.message}`).join(", ")}`
      );
    }
    return responseValidation.data;
  } catch (error) {
    if (error instanceof Error) {
      console.error("Dodo Payments Checkout Session API Error:", {
        message: error.message,
        payload: validation.data,
        config: {
          environment: config.environment,
          hasBearerToken: !!config.bearerToken
        }
      });
      throw new Error(`Failed to create checkout session: ${error.message}`);
    }
    console.error("Unknown error creating checkout session:", error);
    throw new Error(
      "Failed to create checkout session due to an unknown error"
    );
  }
};
var buildCheckoutUrl = async ({
  queryParams,
  body,
  sessionPayload,
  returnUrl,
  bearerToken,
  environment,
  type = "static"
}) => {
  if (type === "session") {
    if (!sessionPayload) {
      throw new Error("sessionPayload is required when type is 'session'");
    }
    const finalPayload = {
      ...sessionPayload,
      return_url: sessionPayload.return_url ?? returnUrl
    };
    const session = await createCheckoutSession(finalPayload, {
      bearerToken,
      environment
    });
    return session.checkout_url;
  }
  const inputData = type === "dynamic" ? body : queryParams;
  let parseResult;
  if (type === "dynamic") {
    parseResult = dynamicCheckoutBodySchema.safeParse(inputData);
  } else {
    parseResult = checkoutQuerySchema.safeParse(inputData);
  }
  const { success, data, error } = parseResult;
  if (!success) {
    throw new Error(
      `Invalid ${type === "dynamic" ? "body" : "query parameters"}.
 ${error.message}`
    );
  }
  if (type !== "dynamic") {
    const {
      productId,
      quantity: quantity2,
      fullName,
      firstName,
      lastName,
      email,
      country,
      addressLine,
      city,
      state,
      zipCode,
      disableFullName,
      disableFirstName,
      disableLastName,
      disableEmail,
      disableCountry,
      disableAddressLine,
      disableCity,
      disableState,
      disableZipCode,
      paymentCurrency,
      showCurrencySelector,
      paymentAmount,
      showDiscounts
      // metadata handled below
    } = data;
    const dodopayments2 = new import_dodopayments.default({
      bearerToken,
      environment
    });
    if (!productId) throw new Error("Missing required field: productId");
    try {
      await dodopayments2.products.retrieve(productId);
    } catch (err) {
      console.error(err);
      throw new Error("Product not found");
    }
    const url = new URL(
      `${environment === "test_mode" ? "https://test.checkout.dodopayments.com" : "https://checkout.dodopayments.com"}/buy/${productId}`
    );
    url.searchParams.set("quantity", quantity2 ? String(quantity2) : "1");
    if (returnUrl) url.searchParams.set("redirect_url", returnUrl);
    if (fullName) url.searchParams.set("fullName", String(fullName));
    if (firstName) url.searchParams.set("firstName", String(firstName));
    if (lastName) url.searchParams.set("lastName", String(lastName));
    if (email) url.searchParams.set("email", String(email));
    if (country) url.searchParams.set("country", String(country));
    if (addressLine) url.searchParams.set("addressLine", String(addressLine));
    if (city) url.searchParams.set("city", String(city));
    if (state) url.searchParams.set("state", String(state));
    if (zipCode) url.searchParams.set("zipCode", String(zipCode));
    if (disableFullName === "true")
      url.searchParams.set("disableFullName", "true");
    if (disableFirstName === "true")
      url.searchParams.set("disableFirstName", "true");
    if (disableLastName === "true")
      url.searchParams.set("disableLastName", "true");
    if (disableEmail === "true") url.searchParams.set("disableEmail", "true");
    if (disableCountry === "true")
      url.searchParams.set("disableCountry", "true");
    if (disableAddressLine === "true")
      url.searchParams.set("disableAddressLine", "true");
    if (disableCity === "true") url.searchParams.set("disableCity", "true");
    if (disableState === "true") url.searchParams.set("disableState", "true");
    if (disableZipCode === "true")
      url.searchParams.set("disableZipCode", "true");
    if (paymentCurrency)
      url.searchParams.set("paymentCurrency", String(paymentCurrency));
    if (showCurrencySelector)
      url.searchParams.set(
        "showCurrencySelector",
        String(showCurrencySelector)
      );
    if (paymentAmount)
      url.searchParams.set("paymentAmount", String(paymentAmount));
    if (showDiscounts)
      url.searchParams.set("showDiscounts", String(showDiscounts));
    for (const [key, value] of Object.entries(queryParams || {})) {
      if (key.startsWith("metadata_") && value && typeof value !== "object") {
        url.searchParams.set(key, String(value));
      }
    }
    return url.toString();
  }
  const dyn = data;
  assertDiscountFieldsExclusive({
    discount_code: dyn.discount_code,
    discount_codes: dyn.discount_codes
  });
  const {
    product_id,
    product_cart,
    quantity,
    billing,
    customer,
    addons,
    metadata,
    allowed_payment_method_types,
    billing_currency,
    discount_code,
    discount_codes,
    on_demand,
    return_url: bodyReturnUrl,
    show_saved_payment_methods,
    tax_id,
    trial_period_days
  } = dyn;
  const dodopayments = new import_dodopayments.default({
    bearerToken,
    environment
  });
  let isSubscription = false;
  let productIdToFetch = product_id;
  if (!product_id && product_cart && product_cart.length > 0) {
    productIdToFetch = product_cart[0].product_id;
  }
  if (!productIdToFetch)
    throw new Error(
      "Missing required field: product_id or product_cart[0].product_id"
    );
  let product;
  try {
    product = await dodopayments.products.retrieve(productIdToFetch);
  } catch (err) {
    console.error(err);
    throw new Error("Product not found");
  }
  isSubscription = Boolean(product.is_recurring);
  if (isSubscription && !product_id)
    throw new Error("Missing required field: product_id for subscription");
  if (!billing) throw new Error("Missing required field: billing");
  if (!customer) throw new Error("Missing required field: customer");
  if (isSubscription) {
    const subscriptionPayload = {
      billing,
      customer,
      product_id,
      quantity: quantity ? Number(quantity) : 1
    };
    if (metadata) subscriptionPayload.metadata = metadata;
    if (discount_codes && discount_codes.length > 0) {
      subscriptionPayload.discount_codes = discount_codes;
    } else if (discount_code) {
      subscriptionPayload.discount_code = discount_code;
    }
    if (addons) subscriptionPayload.addons = addons;
    if (allowed_payment_method_types)
      subscriptionPayload.allowed_payment_method_types = allowed_payment_method_types;
    if (billing_currency)
      subscriptionPayload.billing_currency = billing_currency;
    if (on_demand) subscriptionPayload.on_demand = on_demand;
    subscriptionPayload.payment_link = true;
    if (bodyReturnUrl) {
      subscriptionPayload.return_url = bodyReturnUrl;
    } else if (returnUrl) {
      subscriptionPayload.return_url = returnUrl;
    }
    if (show_saved_payment_methods)
      subscriptionPayload.show_saved_payment_methods = show_saved_payment_methods;
    if (tax_id) subscriptionPayload.tax_id = tax_id;
    if (trial_period_days)
      subscriptionPayload.trial_period_days = trial_period_days;
    let subscription;
    try {
      subscription = await dodopayments.subscriptions.create(subscriptionPayload);
    } catch (err) {
      console.error("Error when creating subscription", err);
      throw new Error(err instanceof Error ? err.message : String(err));
    }
    if (!subscription || !subscription.payment_link) {
      throw new Error(
        "No payment link returned from Dodo Payments API (subscription). Make sure to set payment_link as true in payload"
      );
    }
    return subscription.payment_link;
  } else {
    let cart = product_cart;
    if (!cart && product_id) {
      cart = [
        { product_id, quantity: quantity ? Number(quantity) : 1 }
      ];
    }
    if (!cart || cart.length === 0)
      throw new Error("Missing required field: product_cart or product_id");
    const paymentPayload = {
      billing,
      customer,
      product_cart: cart
    };
    if (metadata) paymentPayload.metadata = metadata;
    paymentPayload.payment_link = true;
    if (allowed_payment_method_types)
      paymentPayload.allowed_payment_method_types = allowed_payment_method_types;
    if (billing_currency) paymentPayload.billing_currency = billing_currency;
    if (discount_codes && discount_codes.length > 0) {
      paymentPayload.discount_codes = discount_codes;
    } else if (discount_code) {
      paymentPayload.discount_code = discount_code;
    }
    if (bodyReturnUrl) {
      paymentPayload.return_url = bodyReturnUrl;
    } else if (returnUrl) {
      paymentPayload.return_url = returnUrl;
    }
    if (show_saved_payment_methods)
      paymentPayload.show_saved_payment_methods = show_saved_payment_methods;
    if (tax_id) paymentPayload.tax_id = tax_id;
    let payment;
    try {
      payment = await dodopayments.payments.create(paymentPayload);
    } catch (err) {
      console.error("Error when creating payment link", err);
      throw new Error(err instanceof Error ? err.message : String(err));
    }
    if (!payment || !payment.payment_link) {
      throw new Error(
        "No payment link returned from Dodo Payments API. Make sure to set payment_link as true in payload."
      );
    }
    return payment.payment_link;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MAX_STACKED_DISCOUNT_CODES,
  assertDiscountFieldsExclusive,
  buildCheckoutUrl,
  checkoutQuerySchema,
  checkoutSessionBillingAddressSchema,
  checkoutSessionCreditEntitlementOverrideSchema,
  checkoutSessionCustomerSchema,
  checkoutSessionCustomizationSchema,
  checkoutSessionFeatureFlagsSchema,
  checkoutSessionOnDemandSchema,
  checkoutSessionPayloadSchema,
  checkoutSessionProductCartItemSchema,
  checkoutSessionResponseSchema,
  checkoutSessionSubscriptionDataSchema,
  createCheckoutSession,
  discountCodesSchema,
  dynamicCheckoutBodySchema,
  paymentMethodTypeSchema
});
