// dodopayments/resources/webhooks/webhooks.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as HeadersAPI from "./headers.mjs";
import { Headers } from "./headers.mjs";
import { Webhook } from 'standardwebhooks';
import { CursorPagePagination } from "../../core/pagination.mjs";
import { buildHeaders } from "../../internal/headers.mjs";
import { path } from "../../internal/utils/path.mjs";
export class Webhooks extends APIResource {
    constructor() {
        super(...arguments);
        this.headers = new HeadersAPI.Headers(this._client);
    }
    /**
     * List all webhooks
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const webhookDetails of client.webhooks.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/webhooks', (CursorPagePagination), { query, ...options });
    }
    /**
     * Create a new webhook
     *
     * @example
     * ```ts
     * const webhookDetails = await client.webhooks.create({
     *   url: 'url',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/webhooks', { body, ...options });
    }
    /**
     * Get a webhook by id
     *
     * @example
     * ```ts
     * const webhookDetails = await client.webhooks.retrieve(
     *   'whk_YdWqVEGKmSYKbsIyDxEab',
     * );
     * ```
     */
    retrieve(webhookID, options) {
        return this._client.get(path `/webhooks/${webhookID}`, options);
    }
    /**
     * Delete a webhook by id
     *
     * @example
     * ```ts
     * await client.webhooks.delete('whk_YdWqVEGKmSYKbsIyDxEab');
     * ```
     */
    delete(webhookID, options) {
        return this._client.delete(path `/webhooks/${webhookID}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Patch a webhook by id
     *
     * @example
     * ```ts
     * const webhookDetails = await client.webhooks.update(
     *   'whk_YdWqVEGKmSYKbsIyDxEab',
     * );
     * ```
     */
    update(webhookID, body, options) {
        return this._client.patch(path `/webhooks/${webhookID}`, { body, ...options });
    }
    /**
     * Get webhook secret by id
     *
     * @example
     * ```ts
     * const response = await client.webhooks.retrieveSecret(
     *   'whk_YdWqVEGKmSYKbsIyDxEab',
     * );
     * ```
     */
    retrieveSecret(webhookID, options) {
        return this._client.get(path `/webhooks/${webhookID}/secret`, options);
    }
    unwrap(body, { headers, key }) {
        if (headers !== undefined) {
            const keyStr = key === undefined ? this._client.webhookKey : key;
            if (keyStr === null)
                throw new Error('Webhook key must not be null in order to unwrap');
            const wh = new Webhook(keyStr);
            wh.verify(body, headers);
        }
        return JSON.parse(body);
    }
    unsafeUnwrap(body) {
        return JSON.parse(body);
    }
}
Webhooks.Headers = Headers;
