// @dodopayments/core/dist/schemas/webhook.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/schemas/webhook.ts
var webhook_exports = {};
__export(webhook_exports, {
  AbandonedCheckoutDetectedPayloadSchema: () => AbandonedCheckoutDetectedPayloadSchema,
  AbandonedCheckoutRecoveredPayloadSchema: () => AbandonedCheckoutRecoveredPayloadSchema,
  AbandonedCheckoutSchema: () => AbandonedCheckoutSchema,
  CreditAddedPayloadSchema: () => CreditAddedPayloadSchema,
  CreditBalanceLowPayloadSchema: () => CreditBalanceLowPayloadSchema,
  CreditBalanceLowSchema: () => CreditBalanceLowSchema,
  CreditDeductedPayloadSchema: () => CreditDeductedPayloadSchema,
  CreditExpiredPayloadSchema: () => CreditExpiredPayloadSchema,
  CreditLedgerEntrySchema: () => CreditLedgerEntrySchema,
  CreditManualAdjustmentPayloadSchema: () => CreditManualAdjustmentPayloadSchema,
  CreditOverageChargedPayloadSchema: () => CreditOverageChargedPayloadSchema,
  CreditRolledOverPayloadSchema: () => CreditRolledOverPayloadSchema,
  CreditRolloverForfeitedPayloadSchema: () => CreditRolloverForfeitedPayloadSchema,
  DisputeAcceptedPayloadSchema: () => DisputeAcceptedPayloadSchema,
  DisputeCancelledPayloadSchema: () => DisputeCancelledPayloadSchema,
  DisputeChallengedPayloadSchema: () => DisputeChallengedPayloadSchema,
  DisputeExpiredPayloadSchema: () => DisputeExpiredPayloadSchema,
  DisputeLostPayloadSchema: () => DisputeLostPayloadSchema,
  DisputeOpenedPayloadSchema: () => DisputeOpenedPayloadSchema,
  DisputeSchema: () => DisputeSchema,
  DisputeWonPayloadSchema: () => DisputeWonPayloadSchema,
  DunningAttemptSchema: () => DunningAttemptSchema,
  DunningRecoveredPayloadSchema: () => DunningRecoveredPayloadSchema,
  DunningStartedPayloadSchema: () => DunningStartedPayloadSchema,
  LicenseKeyCreatedPayloadSchema: () => LicenseKeyCreatedPayloadSchema,
  LicenseKeySchema: () => LicenseKeySchema,
  PaymentCancelledPayloadSchema: () => PaymentCancelledPayloadSchema,
  PaymentFailedPayloadSchema: () => PaymentFailedPayloadSchema,
  PaymentProcessingPayloadSchema: () => PaymentProcessingPayloadSchema,
  PaymentSchema: () => PaymentSchema,
  PaymentSucceededPayloadSchema: () => PaymentSucceededPayloadSchema,
  RefundFailedPayloadSchema: () => RefundFailedPayloadSchema,
  RefundSchema: () => RefundSchema,
  RefundSucceededPayloadSchema: () => RefundSucceededPayloadSchema,
  SubscriptionActivePayloadSchema: () => SubscriptionActivePayloadSchema,
  SubscriptionCancellationScheduledPayloadSchema: () => SubscriptionCancellationScheduledPayloadSchema,
  SubscriptionCancelledPayloadSchema: () => SubscriptionCancelledPayloadSchema,
  SubscriptionExpiredPayloadSchema: () => SubscriptionExpiredPayloadSchema,
  SubscriptionFailedPayloadSchema: () => SubscriptionFailedPayloadSchema,
  SubscriptionOnHoldPayloadSchema: () => SubscriptionOnHoldPayloadSchema,
  SubscriptionPlanChangedPayloadSchema: () => SubscriptionPlanChangedPayloadSchema,
  SubscriptionRenewedPayloadSchema: () => SubscriptionRenewedPayloadSchema,
  SubscriptionSchema: () => SubscriptionSchema,
  SubscriptionTrialEndingPayloadSchema: () => SubscriptionTrialEndingPayloadSchema,
  SubscriptionUpcomingRenewalPayloadSchema: () => SubscriptionUpcomingRenewalPayloadSchema,
  SubscriptionUpdatedPayloadSchema: () => SubscriptionUpdatedPayloadSchema,
  WebhookPayloadSchema: () => WebhookPayloadSchema
});
module.exports = __toCommonJS(webhook_exports);
var import_v3 = require("zod/v3");
var PaymentSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("Payment"),
  billing: import_v3.z.object({
    city: import_v3.z.string().nullable(),
    country: import_v3.z.string(),
    state: import_v3.z.string().nullable(),
    street: import_v3.z.string().nullable(),
    zipcode: import_v3.z.string().nullable()
  }),
  brand_id: import_v3.z.string(),
  business_id: import_v3.z.string(),
  card_holder_name: import_v3.z.string().nullable(),
  card_issuing_country: import_v3.z.string().nullable(),
  card_last_four: import_v3.z.string().nullable(),
  card_network: import_v3.z.string().nullable(),
  card_type: import_v3.z.string().nullable(),
  checkout_session_id: import_v3.z.string().nullable(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  currency: import_v3.z.string(),
  custom_field_responses: import_v3.z.array(
    import_v3.z.object({
      key: import_v3.z.string(),
      value: import_v3.z.string()
    })
  ).nullable(),
  customer: import_v3.z.object({
    customer_id: import_v3.z.string(),
    email: import_v3.z.string(),
    metadata: import_v3.z.record(import_v3.z.any()),
    name: import_v3.z.string(),
    phone_number: import_v3.z.string().nullable()
  }),
  digital_products_delivered: import_v3.z.boolean(),
  /**
   * @deprecated Use `discount_ids` instead. Still populated for backward
   * compatibility with the singular discount_code flow.
   */
  discount_id: import_v3.z.string().nullable(),
  /**
   * Flat list of all stacked discount IDs applied to the payment, in order
   * of application. Always present on v1.98.0+ webhook payloads.
   */
  discount_ids: import_v3.z.array(import_v3.z.string()).nullable().optional(),
  disputes: import_v3.z.array(
    import_v3.z.object({
      amount: import_v3.z.string(),
      business_id: import_v3.z.string(),
      created_at: import_v3.z.string().transform((d) => new Date(d)),
      currency: import_v3.z.string(),
      dispute_id: import_v3.z.string(),
      dispute_stage: import_v3.z.enum(["pre_dispute", "dispute", "pre_arbitration"]),
      dispute_status: import_v3.z.enum([
        "dispute_opened",
        "dispute_expired",
        "dispute_accepted",
        "dispute_cancelled",
        "dispute_challenged",
        "dispute_won",
        "dispute_lost"
      ]),
      payment_id: import_v3.z.string(),
      remarks: import_v3.z.string().nullable()
    })
  ).default([]),
  error_code: import_v3.z.string().nullable(),
  error_message: import_v3.z.string().nullable(),
  invoice_id: import_v3.z.string().nullable(),
  invoice_url: import_v3.z.string().nullable(),
  metadata: import_v3.z.record(import_v3.z.any()),
  payment_id: import_v3.z.string(),
  payment_link: import_v3.z.string().nullable(),
  payment_method: import_v3.z.string().nullable(),
  payment_method_type: import_v3.z.string().nullable(),
  product_cart: import_v3.z.array(
    import_v3.z.object({
      product_id: import_v3.z.string(),
      quantity: import_v3.z.number()
    })
  ).nullable(),
  refunds: import_v3.z.array(
    import_v3.z.object({
      amount: import_v3.z.number().nullable(),
      business_id: import_v3.z.string(),
      created_at: import_v3.z.string().transform((d) => new Date(d)),
      currency: import_v3.z.string().nullable(),
      is_partial: import_v3.z.boolean(),
      payment_id: import_v3.z.string(),
      reason: import_v3.z.string().nullable(),
      refund_id: import_v3.z.string(),
      status: import_v3.z.enum(["succeeded", "failed", "pending", "review"])
    })
  ),
  refund_status: import_v3.z.enum(["partial", "full"]).nullable(),
  settlement_amount: import_v3.z.number(),
  settlement_currency: import_v3.z.string(),
  settlement_tax: import_v3.z.number().nullable(),
  status: import_v3.z.enum([
    "succeeded",
    "failed",
    "cancelled",
    "processing",
    "requires_customer_action",
    "requires_merchant_action",
    "requires_payment_method",
    "requires_confirmation",
    "requires_capture",
    "partially_captured",
    "partially_captured_and_capturable"
  ]).nullable(),
  subscription_id: import_v3.z.string().nullable(),
  tax: import_v3.z.number().nullable(),
  total_amount: import_v3.z.number(),
  updated_at: import_v3.z.string().transform((d) => new Date(d)).nullable()
});
var SubscriptionSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("Subscription"),
  addons: import_v3.z.array(
    import_v3.z.object({
      addon_id: import_v3.z.string(),
      quantity: import_v3.z.number()
    })
  ),
  billing: import_v3.z.object({
    city: import_v3.z.string().nullable(),
    country: import_v3.z.string(),
    state: import_v3.z.string().nullable(),
    street: import_v3.z.string().nullable(),
    zipcode: import_v3.z.string().nullable()
  }),
  cancel_at_next_billing_date: import_v3.z.boolean(),
  cancelled_at: import_v3.z.string().transform((d) => new Date(d)).nullable(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  currency: import_v3.z.string(),
  customer: import_v3.z.object({
    customer_id: import_v3.z.string(),
    email: import_v3.z.string(),
    metadata: import_v3.z.record(import_v3.z.any()),
    name: import_v3.z.string(),
    phone_number: import_v3.z.string().nullable()
  }),
  custom_field_responses: import_v3.z.array(
    import_v3.z.object({
      key: import_v3.z.string(),
      value: import_v3.z.string()
    })
  ).nullable(),
  /**
   * @deprecated Use `discounts[].discount_cycles_remaining` instead. Still
   * populated for backward compatibility with the singular discount_code
   * flow.
   */
  discount_cycles_remaining: import_v3.z.number().nullable(),
  /**
   * @deprecated Use `discounts` (full objects) or `discount_ids` (flat list)
   * instead. Still populated for backward compatibility with the singular
   * discount_code flow.
   */
  discount_id: import_v3.z.string().nullable(),
  /**
   * Flat list of all stacked discount IDs applied to the subscription, in
   * order of application. Always present on v1.98.0+ webhook payloads.
   */
  discount_ids: import_v3.z.array(import_v3.z.string()).nullable().optional(),
  /**
   * All stacked discounts applied to the subscription, in order of
   * application. Each entry contains the discount's position, remaining
   * billing cycles, code, and metadata. Always present on v1.98.0+ webhook
   * payloads.
   */
  discounts: import_v3.z.array(
    import_v3.z.object({
      discount_id: import_v3.z.string(),
      code: import_v3.z.string().nullable().optional(),
      position: import_v3.z.number().int().nonnegative(),
      cycles_remaining: import_v3.z.number().nullable().optional(),
      metadata: import_v3.z.record(import_v3.z.any()).nullable().optional()
    })
  ).nullable().optional(),
  expires_at: import_v3.z.string().transform((d) => new Date(d)).nullable(),
  credit_entitlement_cart: import_v3.z.array(
    import_v3.z.object({
      credit_entitlement_id: import_v3.z.string(),
      credit_entitlement_name: import_v3.z.string(),
      credits_amount: import_v3.z.string(),
      overage_balance: import_v3.z.string(),
      overage_behavior: import_v3.z.enum([
        "forgive_at_reset",
        "invoice_at_billing",
        "carry_deficit",
        "carry_deficit_auto_repay"
      ]),
      overage_enabled: import_v3.z.boolean(),
      product_id: import_v3.z.string(),
      remaining_balance: import_v3.z.string(),
      rollover_enabled: import_v3.z.boolean(),
      unit: import_v3.z.string(),
      expires_after_days: import_v3.z.number().nullable(),
      low_balance_threshold_percent: import_v3.z.number().nullable(),
      max_rollover_count: import_v3.z.number().nullable(),
      overage_limit: import_v3.z.string().nullable(),
      rollover_percentage: import_v3.z.number().nullable(),
      rollover_timeframe_count: import_v3.z.number().nullable(),
      rollover_timeframe_interval: import_v3.z.enum(["Day", "Week", "Month", "Year"]).nullable()
    })
  ),
  meter_credit_entitlement_cart: import_v3.z.array(
    import_v3.z.object({
      credit_entitlement_id: import_v3.z.string(),
      meter_id: import_v3.z.string(),
      meter_name: import_v3.z.string(),
      meter_units_per_credit: import_v3.z.string(),
      product_id: import_v3.z.string()
    })
  ),
  meters: import_v3.z.array(
    import_v3.z.object({
      currency: import_v3.z.string(),
      description: import_v3.z.string().nullable(),
      free_threshold: import_v3.z.number(),
      measurement_unit: import_v3.z.string(),
      meter_id: import_v3.z.string(),
      name: import_v3.z.string(),
      price_per_unit: import_v3.z.string().nullable()
    })
  ),
  metadata: import_v3.z.record(import_v3.z.any()),
  next_billing_date: import_v3.z.string().transform((d) => new Date(d)),
  on_demand: import_v3.z.boolean(),
  payment_frequency_count: import_v3.z.number(),
  payment_frequency_interval: import_v3.z.enum(["Day", "Week", "Month", "Year"]),
  payment_method_id: import_v3.z.string().nullable(),
  previous_billing_date: import_v3.z.string().transform((d) => new Date(d)),
  product_id: import_v3.z.string(),
  quantity: import_v3.z.number(),
  recurring_pre_tax_amount: import_v3.z.number(),
  status: import_v3.z.enum([
    "pending",
    "active",
    "on_hold",
    "cancelled",
    "expired",
    "failed"
  ]),
  subscription_id: import_v3.z.string(),
  subscription_period_count: import_v3.z.number(),
  subscription_period_interval: import_v3.z.enum(["Day", "Week", "Month", "Year"]),
  tax_id: import_v3.z.string().nullable(),
  tax_inclusive: import_v3.z.boolean(),
  trial_period_days: import_v3.z.number()
});
var RefundSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("Refund"),
  amount: import_v3.z.number().nullable(),
  business_id: import_v3.z.string(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  customer: import_v3.z.object({
    customer_id: import_v3.z.string(),
    email: import_v3.z.string(),
    metadata: import_v3.z.record(import_v3.z.any()),
    name: import_v3.z.string(),
    phone_number: import_v3.z.string().nullable()
  }),
  currency: import_v3.z.string().nullable(),
  is_partial: import_v3.z.boolean(),
  metadata: import_v3.z.record(import_v3.z.any()),
  payment_id: import_v3.z.string(),
  reason: import_v3.z.string().nullable(),
  refund_id: import_v3.z.string(),
  status: import_v3.z.enum(["succeeded", "failed", "pending", "review"])
});
var DisputeSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("Dispute"),
  amount: import_v3.z.string(),
  business_id: import_v3.z.string(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  currency: import_v3.z.string(),
  customer: import_v3.z.object({
    customer_id: import_v3.z.string(),
    email: import_v3.z.string(),
    metadata: import_v3.z.record(import_v3.z.any()),
    name: import_v3.z.string(),
    phone_number: import_v3.z.string().nullable()
  }),
  dispute_id: import_v3.z.string(),
  dispute_stage: import_v3.z.enum(["pre_dispute", "dispute", "pre_arbitration"]),
  dispute_status: import_v3.z.enum([
    "dispute_opened",
    "dispute_expired",
    "dispute_accepted",
    "dispute_cancelled",
    "dispute_challenged",
    "dispute_won",
    "dispute_lost"
  ]),
  payment_id: import_v3.z.string(),
  reason: import_v3.z.string().nullable(),
  remarks: import_v3.z.string().nullable()
});
var LicenseKeySchema = import_v3.z.object({
  payload_type: import_v3.z.literal("LicenseKey"),
  activations_limit: import_v3.z.number().nullable(),
  business_id: import_v3.z.string(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  customer_id: import_v3.z.string(),
  expires_at: import_v3.z.string().transform((d) => new Date(d)).nullable(),
  id: import_v3.z.string(),
  instances_count: import_v3.z.number(),
  key: import_v3.z.string(),
  payment_id: import_v3.z.string(),
  product_id: import_v3.z.string(),
  status: import_v3.z.enum(["active", "expired", "disabled"]),
  subscription_id: import_v3.z.string().nullable()
});
var PaymentSucceededPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("payment.succeeded"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: PaymentSchema
});
var PaymentFailedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("payment.failed"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: PaymentSchema
});
var PaymentProcessingPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("payment.processing"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: PaymentSchema
});
var PaymentCancelledPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("payment.cancelled"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: PaymentSchema
});
var RefundSucceededPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("refund.succeeded"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: RefundSchema
});
var RefundFailedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("refund.failed"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: RefundSchema
});
var DisputeOpenedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.opened"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeExpiredPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.expired"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeAcceptedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.accepted"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeCancelledPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.cancelled"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeChallengedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.challenged"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeWonPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.won"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var DisputeLostPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dispute.lost"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DisputeSchema
});
var SubscriptionActivePayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.active"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionOnHoldPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.on_hold"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionRenewedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.renewed"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionPlanChangedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.plan_changed"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionCancelledPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.cancelled"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionCancellationScheduledPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.cancellation_scheduled"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionTrialEndingPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.trial_ending"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionUpcomingRenewalPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.upcoming_renewal"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionFailedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.failed"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionExpiredPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.expired"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var SubscriptionUpdatedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("subscription.updated"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: SubscriptionSchema
});
var LicenseKeyCreatedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("license_key.created"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: LicenseKeySchema
});
var CreditLedgerEntrySchema = import_v3.z.object({
  payload_type: import_v3.z.literal("CreditLedgerEntry"),
  id: import_v3.z.string(),
  amount: import_v3.z.string(),
  balance_after: import_v3.z.string(),
  balance_before: import_v3.z.string(),
  business_id: import_v3.z.string(),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  credit_entitlement_id: import_v3.z.string(),
  customer_id: import_v3.z.string(),
  is_credit: import_v3.z.boolean(),
  overage_after: import_v3.z.string(),
  overage_before: import_v3.z.string(),
  transaction_type: import_v3.z.enum([
    "credit_added",
    "credit_deducted",
    "credit_expired",
    "credit_rolled_over",
    "rollover_forfeited",
    "overage_charged",
    "auto_top_up",
    "manual_adjustment",
    "refund"
  ]),
  description: import_v3.z.string().nullable(),
  grant_id: import_v3.z.string().nullable(),
  reference_id: import_v3.z.string().nullable(),
  reference_type: import_v3.z.string().nullable()
});
var CreditBalanceLowSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("CreditBalanceLow"),
  customer_id: import_v3.z.string(),
  subscription_id: import_v3.z.string(),
  credit_entitlement_id: import_v3.z.string(),
  credit_entitlement_name: import_v3.z.string(),
  available_balance: import_v3.z.string(),
  subscription_credits_amount: import_v3.z.string(),
  threshold_percent: import_v3.z.number(),
  threshold_amount: import_v3.z.string()
});
var CreditAddedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.added"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditDeductedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.deducted"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditExpiredPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.expired"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditRolledOverPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.rolled_over"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditRolloverForfeitedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.rollover_forfeited"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditOverageChargedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.overage_charged"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditManualAdjustmentPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.manual_adjustment"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditLedgerEntrySchema
});
var CreditBalanceLowPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("credit.balance_low"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: CreditBalanceLowSchema
});
var AbandonedCheckoutSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("AbandonedCheckout"),
  abandoned_at: import_v3.z.string().transform((d) => new Date(d)),
  abandonment_reason: import_v3.z.enum(["payment_failed", "checkout_incomplete"]),
  customer_id: import_v3.z.string(),
  payment_id: import_v3.z.string(),
  status: import_v3.z.enum(["abandoned", "recovering", "recovered", "exhausted", "opted_out"]),
  recovered_payment_id: import_v3.z.string().nullable().optional()
});
var DunningAttemptSchema = import_v3.z.object({
  payload_type: import_v3.z.literal("DunningAttempt"),
  created_at: import_v3.z.string().transform((d) => new Date(d)),
  customer_id: import_v3.z.string(),
  status: import_v3.z.enum(["recovering", "recovered", "exhausted"]),
  subscription_id: import_v3.z.string(),
  trigger_state: import_v3.z.enum(["on_hold", "cancelled"]),
  payment_id: import_v3.z.string().nullable().optional()
});
var AbandonedCheckoutDetectedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("abandoned_checkout.detected"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: AbandonedCheckoutSchema
});
var AbandonedCheckoutRecoveredPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("abandoned_checkout.recovered"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: AbandonedCheckoutSchema
});
var DunningStartedPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dunning.started"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DunningAttemptSchema
});
var DunningRecoveredPayloadSchema = import_v3.z.object({
  business_id: import_v3.z.string(),
  type: import_v3.z.literal("dunning.recovered"),
  timestamp: import_v3.z.string().transform((d) => new Date(d)),
  data: DunningAttemptSchema
});
var WebhookPayloadSchema = import_v3.z.discriminatedUnion("type", [
  PaymentSucceededPayloadSchema,
  PaymentFailedPayloadSchema,
  PaymentProcessingPayloadSchema,
  PaymentCancelledPayloadSchema,
  RefundSucceededPayloadSchema,
  RefundFailedPayloadSchema,
  DisputeOpenedPayloadSchema,
  DisputeExpiredPayloadSchema,
  DisputeAcceptedPayloadSchema,
  DisputeCancelledPayloadSchema,
  DisputeChallengedPayloadSchema,
  DisputeWonPayloadSchema,
  DisputeLostPayloadSchema,
  SubscriptionActivePayloadSchema,
  SubscriptionOnHoldPayloadSchema,
  SubscriptionRenewedPayloadSchema,
  SubscriptionPlanChangedPayloadSchema,
  SubscriptionCancelledPayloadSchema,
  SubscriptionCancellationScheduledPayloadSchema,
  SubscriptionTrialEndingPayloadSchema,
  SubscriptionUpcomingRenewalPayloadSchema,
  SubscriptionFailedPayloadSchema,
  SubscriptionExpiredPayloadSchema,
  SubscriptionUpdatedPayloadSchema,
  LicenseKeyCreatedPayloadSchema,
  AbandonedCheckoutDetectedPayloadSchema,
  AbandonedCheckoutRecoveredPayloadSchema,
  DunningStartedPayloadSchema,
  DunningRecoveredPayloadSchema,
  CreditAddedPayloadSchema,
  CreditDeductedPayloadSchema,
  CreditExpiredPayloadSchema,
  CreditRolledOverPayloadSchema,
  CreditRolloverForfeitedPayloadSchema,
  CreditOverageChargedPayloadSchema,
  CreditManualAdjustmentPayloadSchema,
  CreditBalanceLowPayloadSchema
]);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AbandonedCheckoutDetectedPayloadSchema,
  AbandonedCheckoutRecoveredPayloadSchema,
  AbandonedCheckoutSchema,
  CreditAddedPayloadSchema,
  CreditBalanceLowPayloadSchema,
  CreditBalanceLowSchema,
  CreditDeductedPayloadSchema,
  CreditExpiredPayloadSchema,
  CreditLedgerEntrySchema,
  CreditManualAdjustmentPayloadSchema,
  CreditOverageChargedPayloadSchema,
  CreditRolledOverPayloadSchema,
  CreditRolloverForfeitedPayloadSchema,
  DisputeAcceptedPayloadSchema,
  DisputeCancelledPayloadSchema,
  DisputeChallengedPayloadSchema,
  DisputeExpiredPayloadSchema,
  DisputeLostPayloadSchema,
  DisputeOpenedPayloadSchema,
  DisputeSchema,
  DisputeWonPayloadSchema,
  DunningAttemptSchema,
  DunningRecoveredPayloadSchema,
  DunningStartedPayloadSchema,
  LicenseKeyCreatedPayloadSchema,
  LicenseKeySchema,
  PaymentCancelledPayloadSchema,
  PaymentFailedPayloadSchema,
  PaymentProcessingPayloadSchema,
  PaymentSchema,
  PaymentSucceededPayloadSchema,
  RefundFailedPayloadSchema,
  RefundSchema,
  RefundSucceededPayloadSchema,
  SubscriptionActivePayloadSchema,
  SubscriptionCancellationScheduledPayloadSchema,
  SubscriptionCancelledPayloadSchema,
  SubscriptionExpiredPayloadSchema,
  SubscriptionFailedPayloadSchema,
  SubscriptionOnHoldPayloadSchema,
  SubscriptionPlanChangedPayloadSchema,
  SubscriptionRenewedPayloadSchema,
  SubscriptionSchema,
  SubscriptionTrialEndingPayloadSchema,
  SubscriptionUpcomingRenewalPayloadSchema,
  SubscriptionUpdatedPayloadSchema,
  WebhookPayloadSchema
});
