// dodopayments/resources/product-collections/product-collections.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductCollections = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const GroupsAPI = tslib_1.__importStar(require("./groups/groups.js"));
const groups_1 = require("./groups/groups.js");
const pagination_1 = require("../../core/pagination.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class ProductCollections extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.groups = new GroupsAPI.Groups(this._client);
    }
    /**
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const productCollectionListResponse of client.productCollections.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/product-collections', (pagination_1.DefaultPageNumberPagination), { query, ...options });
    }
    /**
     * @example
     * ```ts
     * const productCollection =
     *   await client.productCollections.create({
     *     groups: [{ products: [{ product_id: 'product_id' }] }],
     *     name: 'name',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/product-collections', { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const productCollection =
     *   await client.productCollections.retrieve(
     *     'pdc_8BWv0hojwUH7iCDabr0NI',
     *   );
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/product-collections/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * await client.productCollections.delete(
     *   'pdc_8BWv0hojwUH7iCDabr0NI',
     * );
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/product-collections/${id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * await client.productCollections.update(
     *   'pdc_8BWv0hojwUH7iCDabr0NI',
     * );
     * ```
     */
    update(id, body, options) {
        return this._client.patch((0, path_1.path) `/product-collections/${id}`, {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * const response =
     *   await client.productCollections.updateImages(
     *     'pdc_8BWv0hojwUH7iCDabr0NI',
     *   );
     * ```
     */
    updateImages(id, params = {}, options) {
        const { force_update } = params ?? {};
        return this._client.put((0, path_1.path) `/product-collections/${id}/images`, { query: { force_update }, ...options });
    }
    /**
     * @example
     * ```ts
     * const response = await client.productCollections.unarchive(
     *   'pdc_8BWv0hojwUH7iCDabr0NI',
     * );
     * ```
     */
    unarchive(id, options) {
        return this._client.post((0, path_1.path) `/product-collections/${id}/unarchive`, options);
    }
}
exports.ProductCollections = ProductCollections;
ProductCollections.Groups = groups_1.Groups;
