// dodopayments/resources/product-collections/product-collections.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as GroupsAPI from "./groups/groups.mjs";
import { Groups, } from "./groups/groups.mjs";
import { DefaultPageNumberPagination, } from "../../core/pagination.mjs";
import { buildHeaders } from "../../internal/headers.mjs";
import { path } from "../../internal/utils/path.mjs";
export class ProductCollections extends APIResource {
    constructor() {
        super(...arguments);
        this.groups = new GroupsAPI.Groups(this._client);
    }
    /**
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const productCollectionListResponse of client.productCollections.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/product-collections', (DefaultPageNumberPagination), { query, ...options });
    }
    /**
     * @example
     * ```ts
     * const productCollection =
     *   await client.productCollections.create({
     *     groups: [{ products: [{ product_id: 'product_id' }] }],
     *     name: 'name',
     *   });
     * ```
     */
    create(body, options) {
        return this._client.post('/product-collections', { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const productCollection =
     *   await client.productCollections.retrieve(
     *     'pdc_8BWv0hojwUH7iCDabr0NI',
     *   );
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/product-collections/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * await client.productCollections.delete(
     *   'pdc_8BWv0hojwUH7iCDabr0NI',
     * );
     * ```
     */
    delete(id, options) {
        return this._client.delete(path `/product-collections/${id}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * await client.productCollections.update(
     *   'pdc_8BWv0hojwUH7iCDabr0NI',
     * );
     * ```
     */
    update(id, body, options) {
        return this._client.patch(path `/product-collections/${id}`, {
            body,
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * const response =
     *   await client.productCollections.updateImages(
     *     'pdc_8BWv0hojwUH7iCDabr0NI',
     *   );
     * ```
     */
    updateImages(id, params = {}, options) {
        const { force_update } = params ?? {};
        return this._client.put(path `/product-collections/${id}/images`, { query: { force_update }, ...options });
    }
    /**
     * @example
     * ```ts
     * const response = await client.productCollections.unarchive(
     *   'pdc_8BWv0hojwUH7iCDabr0NI',
     * );
     * ```
     */
    unarchive(id, options) {
        return this._client.post(path `/product-collections/${id}/unarchive`, options);
    }
}
ProductCollections.Groups = Groups;
