// dodopayments/resources/addons.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { DefaultPageNumberPagination, } from "../core/pagination.mjs";
import { path } from "../internal/utils/path.mjs";
export class Addons extends APIResource {
    /**
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const addonResponse of client.addons.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/addons', (DefaultPageNumberPagination), {
            query,
            ...options,
        });
    }
    /**
     * @example
     * ```ts
     * const addonResponse = await client.addons.create({
     *   currency: 'AED',
     *   name: 'name',
     *   price: 0,
     *   tax_category: 'digital_products',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/addons', { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const addonResponse = await client.addons.retrieve(
     *   'adn_NX1zdqW4Hbivsqz8vI9dc',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/addons/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * const addonResponse = await client.addons.update(
     *   'adn_NX1zdqW4Hbivsqz8vI9dc',
     * );
     * ```
     */
    update(id, body, options) {
        return this._client.patch(path `/addons/${id}`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const response = await client.addons.updateImages(
     *   'adn_NX1zdqW4Hbivsqz8vI9dc',
     * );
     * ```
     */
    updateImages(id, options) {
        return this._client.put(path `/addons/${id}/images`, options);
    }
}
