// dodopayments/resources/addons.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Addons = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const path_1 = require("../internal/utils/path.js");
class Addons extends resource_1.APIResource {
    /**
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const addonResponse of client.addons.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/addons', (pagination_1.DefaultPageNumberPagination), {
            query,
            ...options,
        });
    }
    /**
     * @example
     * ```ts
     * const addonResponse = await client.addons.create({
     *   currency: 'AED',
     *   name: 'name',
     *   price: 0,
     *   tax_category: 'digital_products',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/addons', { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const addonResponse = await client.addons.retrieve(
     *   'adn_NX1zdqW4Hbivsqz8vI9dc',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get((0, path_1.path) `/addons/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * const addonResponse = await client.addons.update(
     *   'adn_NX1zdqW4Hbivsqz8vI9dc',
     * );
     * ```
     */
    update(id, body, options) {
        return this._client.patch((0, path_1.path) `/addons/${id}`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const response = await client.addons.updateImages(
     *   'adn_NX1zdqW4Hbivsqz8vI9dc',
     * );
     * ```
     */
    updateImages(id, options) {
        return this._client.put((0, path_1.path) `/addons/${id}/images`, options);
    }
}
exports.Addons = Addons;
