// dodopayments/resources/brands.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { path } from "../internal/utils/path.mjs";
export class Brands extends APIResource {
    /**
     * @example
     * ```ts
     * const brands = await client.brands.list();
     * ```
     */
    list(options) {
        return this._client.get('/brands', options);
    }
    /**
     * @example
     * ```ts
     * const brand = await client.brands.create();
     * ```
     */
    create(body, options) {
        return this._client.post('/brands', { body, ...options });
    }
    /**
     * Thin handler just calls `get_brand` and wraps in `Json(...)`
     *
     * @example
     * ```ts
     * const brand = await client.brands.retrieve(
     *   'brnd_8dFiAW42v28JzhlVSocjq',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/brands/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * const brand = await client.brands.update(
     *   'brnd_8dFiAW42v28JzhlVSocjq',
     * );
     * ```
     */
    update(id, body, options) {
        return this._client.patch(path `/brands/${id}`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const response = await client.brands.updateImages(
     *   'brnd_8dFiAW42v28JzhlVSocjq',
     * );
     * ```
     */
    updateImages(id, options) {
        return this._client.put(path `/brands/${id}/images`, options);
    }
}
