// dodopayments/resources/discounts.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { DefaultPageNumberPagination, } from "../core/pagination.mjs";
import { buildHeaders } from "../internal/headers.mjs";
import { path } from "../internal/utils/path.mjs";
export class Discounts extends APIResource {
    /**
     * GET /discounts
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const discount of client.discounts.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/discounts', (DefaultPageNumberPagination), {
            query,
            ...options,
        });
    }
    /**
     * POST /discounts If `code` is omitted or empty, a random 16-char uppercase code
     * is generated.
     *
     * @example
     * ```ts
     * const discount = await client.discounts.create({
     *   amount: 0,
     *   type: 'percentage',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/discounts', { body, ...options });
    }
    /**
     * GET /discounts/{discount_id}
     *
     * @example
     * ```ts
     * const discount = await client.discounts.retrieve(
     *   'dsc_qxxEmg5PuM1uNTE0LgkP9',
     * );
     * ```
     */
    retrieve(discountID, options) {
        return this._client.get(path `/discounts/${discountID}`, options);
    }
    /**
     * DELETE /discounts/{discount_id}
     *
     * @example
     * ```ts
     * await client.discounts.delete('dsc_qxxEmg5PuM1uNTE0LgkP9');
     * ```
     */
    delete(discountID, options) {
        return this._client.delete(path `/discounts/${discountID}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * PATCH /discounts/{discount_id}
     *
     * @example
     * ```ts
     * const discount = await client.discounts.update(
     *   'dsc_qxxEmg5PuM1uNTE0LgkP9',
     * );
     * ```
     */
    update(discountID, body, options) {
        return this._client.patch(path `/discounts/${discountID}`, { body, ...options });
    }
    /**
     * Validate and fetch a discount by its code name (e.g., "SAVE20"). This allows
     * real-time validation directly against the API using the human-readable discount
     * code instead of requiring the internal discount_id.
     *
     * @example
     * ```ts
     * const discount = await client.discounts.retrieveByCode(
     *   'code',
     * );
     * ```
     */
    retrieveByCode(code, options) {
        return this._client.get(path `/discounts/code/${code}`, options);
    }
}
