// dodopayments/resources/discounts.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Discounts = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const headers_1 = require("../internal/headers.js");
const path_1 = require("../internal/utils/path.js");
class Discounts extends resource_1.APIResource {
    /**
     * GET /discounts
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const discount of client.discounts.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/discounts', (pagination_1.DefaultPageNumberPagination), {
            query,
            ...options,
        });
    }
    /**
     * POST /discounts If `code` is omitted or empty, a random 16-char uppercase code
     * is generated.
     *
     * @example
     * ```ts
     * const discount = await client.discounts.create({
     *   amount: 0,
     *   type: 'percentage',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/discounts', { body, ...options });
    }
    /**
     * GET /discounts/{discount_id}
     *
     * @example
     * ```ts
     * const discount = await client.discounts.retrieve(
     *   'dsc_qxxEmg5PuM1uNTE0LgkP9',
     * );
     * ```
     */
    retrieve(discountID, options) {
        return this._client.get((0, path_1.path) `/discounts/${discountID}`, options);
    }
    /**
     * DELETE /discounts/{discount_id}
     *
     * @example
     * ```ts
     * await client.discounts.delete('dsc_qxxEmg5PuM1uNTE0LgkP9');
     * ```
     */
    delete(discountID, options) {
        return this._client.delete((0, path_1.path) `/discounts/${discountID}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * PATCH /discounts/{discount_id}
     *
     * @example
     * ```ts
     * const discount = await client.discounts.update(
     *   'dsc_qxxEmg5PuM1uNTE0LgkP9',
     * );
     * ```
     */
    update(discountID, body, options) {
        return this._client.patch((0, path_1.path) `/discounts/${discountID}`, { body, ...options });
    }
    /**
     * Validate and fetch a discount by its code name (e.g., "SAVE20"). This allows
     * real-time validation directly against the API using the human-readable discount
     * code instead of requiring the internal discount_id.
     *
     * @example
     * ```ts
     * const discount = await client.discounts.retrieveByCode(
     *   'code',
     * );
     * ```
     */
    retrieveByCode(code, options) {
        return this._client.get((0, path_1.path) `/discounts/code/${code}`, options);
    }
}
exports.Discounts = Discounts;
