// dodopayments/resources/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhooks = exports.WebhookEvents = exports.UsageEvents = exports.Subscriptions = exports.Refunds = exports.Products = exports.ProductCollections = exports.Payouts = exports.Payments = exports.Misc = exports.Meters = exports.Licenses = exports.LicenseKeys = exports.LicenseKeyInstances = exports.Invoices = exports.Entitlements = exports.Disputes = exports.Discounts = exports.Customers = exports.CreditEntitlements = exports.CheckoutSessions = exports.Brands = exports.Balances = exports.Addons = void 0;
var addons_1 = require("./addons.js");
Object.defineProperty(exports, "Addons", { enumerable: true, get: function () { return addons_1.Addons; } });
var balances_1 = require("./balances.js");
Object.defineProperty(exports, "Balances", { enumerable: true, get: function () { return balances_1.Balances; } });
var brands_1 = require("./brands.js");
Object.defineProperty(exports, "Brands", { enumerable: true, get: function () { return brands_1.Brands; } });
var checkout_sessions_1 = require("./checkout-sessions.js");
Object.defineProperty(exports, "CheckoutSessions", { enumerable: true, get: function () { return checkout_sessions_1.CheckoutSessions; } });
var credit_entitlements_1 = require("./credit-entitlements/credit-entitlements.js");
Object.defineProperty(exports, "CreditEntitlements", { enumerable: true, get: function () { return credit_entitlements_1.CreditEntitlements; } });
var customers_1 = require("./customers/customers.js");
Object.defineProperty(exports, "Customers", { enumerable: true, get: function () { return customers_1.Customers; } });
var discounts_1 = require("./discounts.js");
Object.defineProperty(exports, "Discounts", { enumerable: true, get: function () { return discounts_1.Discounts; } });
var disputes_1 = require("./disputes.js");
Object.defineProperty(exports, "Disputes", { enumerable: true, get: function () { return disputes_1.Disputes; } });
var entitlements_1 = require("./entitlements/entitlements.js");
Object.defineProperty(exports, "Entitlements", { enumerable: true, get: function () { return entitlements_1.Entitlements; } });
var invoices_1 = require("./invoices/invoices.js");
Object.defineProperty(exports, "Invoices", { enumerable: true, get: function () { return invoices_1.Invoices; } });
var license_key_instances_1 = require("./license-key-instances.js");
Object.defineProperty(exports, "LicenseKeyInstances", { enumerable: true, get: function () { return license_key_instances_1.LicenseKeyInstances; } });
var license_keys_1 = require("./license-keys.js");
Object.defineProperty(exports, "LicenseKeys", { enumerable: true, get: function () { return license_keys_1.LicenseKeys; } });
var licenses_1 = require("./licenses.js");
Object.defineProperty(exports, "Licenses", { enumerable: true, get: function () { return licenses_1.Licenses; } });
var meters_1 = require("./meters.js");
Object.defineProperty(exports, "Meters", { enumerable: true, get: function () { return meters_1.Meters; } });
var misc_1 = require("./misc.js");
Object.defineProperty(exports, "Misc", { enumerable: true, get: function () { return misc_1.Misc; } });
var payments_1 = require("./payments.js");
Object.defineProperty(exports, "Payments", { enumerable: true, get: function () { return payments_1.Payments; } });
var payouts_1 = require("./payouts/payouts.js");
Object.defineProperty(exports, "Payouts", { enumerable: true, get: function () { return payouts_1.Payouts; } });
var product_collections_1 = require("./product-collections/product-collections.js");
Object.defineProperty(exports, "ProductCollections", { enumerable: true, get: function () { return product_collections_1.ProductCollections; } });
var products_1 = require("./products/products.js");
Object.defineProperty(exports, "Products", { enumerable: true, get: function () { return products_1.Products; } });
var refunds_1 = require("./refunds.js");
Object.defineProperty(exports, "Refunds", { enumerable: true, get: function () { return refunds_1.Refunds; } });
var subscriptions_1 = require("./subscriptions.js");
Object.defineProperty(exports, "Subscriptions", { enumerable: true, get: function () { return subscriptions_1.Subscriptions; } });
var usage_events_1 = require("./usage-events.js");
Object.defineProperty(exports, "UsageEvents", { enumerable: true, get: function () { return usage_events_1.UsageEvents; } });
var webhook_events_1 = require("./webhook-events.js");
Object.defineProperty(exports, "WebhookEvents", { enumerable: true, get: function () { return webhook_events_1.WebhookEvents; } });
var webhooks_1 = require("./webhooks/webhooks.js");
Object.defineProperty(exports, "Webhooks", { enumerable: true, get: function () { return webhooks_1.Webhooks; } });
