// dodopayments/resources/meters.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { DefaultPageNumberPagination, } from "../core/pagination.mjs";
import { buildHeaders } from "../internal/headers.mjs";
import { path } from "../internal/utils/path.mjs";
export class Meters extends APIResource {
    /**
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const meter of client.meters.list()) {
     *   // ...
     * }
     * ```
     */
    list(query = {}, options) {
        return this._client.getAPIList('/meters', (DefaultPageNumberPagination), { query, ...options });
    }
    /**
     * @example
     * ```ts
     * const meter = await client.meters.create({
     *   aggregation: { type: 'count' },
     *   event_name: 'event_name',
     *   measurement_unit: 'measurement_unit',
     *   name: 'name',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/meters', { body, ...options });
    }
    /**
     * @example
     * ```ts
     * const meter = await client.meters.retrieve(
     *   'mtr_h5tgTWL55OyMO0L2Q9w9v',
     * );
     * ```
     */
    retrieve(id, options) {
        return this._client.get(path `/meters/${id}`, options);
    }
    /**
     * @example
     * ```ts
     * await client.meters.archive('mtr_h5tgTWL55OyMO0L2Q9w9v');
     * ```
     */
    archive(id, options) {
        return this._client.delete(path `/meters/${id}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * await client.meters.unarchive('mtr_h5tgTWL55OyMO0L2Q9w9v');
     * ```
     */
    unarchive(id, options) {
        return this._client.post(path `/meters/${id}/unarchive`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
