// svix/dist/api_internal/messagePollerv2.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessagePollerv2 = void 0;
const pollerV2CommitIn_1 = require("../models/pollerV2CommitIn");
const pollerV2PollOut_1 = require("../models/pollerV2PollOut");
const request_1 = require("../request");
class MessagePollerv2 {
    constructor(requestCtx) {
        this.requestCtx = requestCtx;
    }
    consumerPoll(appId, sinkId, consumerId, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.GET, "/api/v1/app/{app_id}/polling-endpoint/{sink_id}/consumer/{consumer_id}");
            request.setPathParam("app_id", appId);
            request.setPathParam("sink_id", sinkId);
            request.setPathParam("consumer_id", consumerId);
            request.setQueryParams({
                limit: options === null || options === void 0 ? void 0 : options.limit,
                lease_duration_ms: options === null || options === void 0 ? void 0 : options.leaseDurationMs,
                starting_position: options === null || options === void 0 ? void 0 : options.startingPosition,
            });
            return yield request.send(this.requestCtx, pollerV2PollOut_1.PollerV2PollOutSerializer._fromJsonObject);
        });
    }
    consumerCommit(appId, sinkId, consumerId, pollerV2CommitIn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const request = new request_1.SvixRequest(request_1.HttpMethod.POST, "/api/v1/app/{app_id}/polling-endpoint/{sink_id}/consumer/{consumer_id}/commit");
            request.setPathParam("app_id", appId);
            request.setPathParam("sink_id", sinkId);
            request.setPathParam("consumer_id", consumerId);
            request.setHeaderParam("idempotency-key", options === null || options === void 0 ? void 0 : options.idempotencyKey);
            request.setBody(pollerV2CommitIn_1.PollerV2CommitInSerializer._toJsonObject(pollerV2CommitIn));
            return yield request.sendNoResponseBody(this.requestCtx);
        });
    }
}
exports.MessagePollerv2 = MessagePollerv2;
