// @clerk/shared/dist/utils/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_runtimeEnvironment = require('../_chunks/runtimeEnvironment-DVXDje4R.js');
const require_handleValueOrFn = require('../_chunks/handleValueOrFn-CCSXb9Ik.js');
const require_instance = require('../_chunks/instance-DcOvlnfl.js');
const require_noop = require('../_chunks/noop-DCQKAlxF.js');
const require_createDeferredPromise = require('../_chunks/createDeferredPromise-BXa3LUQE.js');
const require_utils = require('../_chunks/utils-DccuyyQf.js');

exports.allSettled = require_utils.allSettled;
exports.automatedEnvironmentVariables = require_runtimeEnvironment.automatedEnvironmentVariables;
exports.createDeferredPromise = require_createDeferredPromise.createDeferredPromise;
exports.fastDeepMergeAndKeep = require_utils.fastDeepMergeAndKeep;
exports.fastDeepMergeAndReplace = require_utils.fastDeepMergeAndReplace;
exports.handleValueOrFn = require_handleValueOrFn.handleValueOrFn;
exports.isAutomatedEnvironment = require_runtimeEnvironment.isAutomatedEnvironment;
exports.isDevelopmentEnvironment = require_runtimeEnvironment.isDevelopmentEnvironment;
exports.isProductionEnvironment = require_runtimeEnvironment.isProductionEnvironment;
exports.isStaging = require_instance.isStaging;
exports.isTestEnvironment = require_runtimeEnvironment.isTestEnvironment;
exports.logErrorInDevMode = require_utils.logErrorInDevMode;
exports.noop = require_noop.noop;
exports.runIfFunctionOrReturn = require_utils.runIfFunctionOrReturn;
