// svix/dist/autoconfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoConfig = exports.decodeAutoconfigTokenV1 = exports.isAutoConfigTokenContentV1 = exports.AutoConfigError = void 0;
const api_internal_1 = require("./api_internal");
const endpoint_1 = require("./api_internal/endpoint");
const webhook_1 = require("./webhook");
const AUTOCONFIG_TOKEN_PREFIX_V1 = "auto_v1_";
class AutoConfigError extends Error {
    constructor(message = "invalid token") {
        super(message);
        this.name = "AutoConfigError";
    }
}
exports.AutoConfigError = AutoConfigError;
function isAutoConfigTokenContentV1(value) {
    if (typeof value !== "object" || value === null) {
        return false;
    }
    const { aid, eid, surl, esec, tok } = value;
    return (typeof aid === "string" &&
        typeof eid === "string" &&
        typeof surl === "string" &&
        typeof esec === "string" &&
        typeof tok === "string");
}
exports.isAutoConfigTokenContentV1 = isAutoConfigTokenContentV1;
function decodeAutoconfigTokenV1(token) {
    if (!token.startsWith(AUTOCONFIG_TOKEN_PREFIX_V1)) {
        throw new AutoConfigError();
    }
    const b64 = token.slice(AUTOCONFIG_TOKEN_PREFIX_V1.length);
    let json;
    try {
        json = Buffer.from(b64, "base64").toString("utf8");
    }
    catch (_a) {
        throw new AutoConfigError();
    }
    let parsed;
    try {
        parsed = JSON.parse(json);
    }
    catch (_b) {
        throw new AutoConfigError();
    }
    if (!isAutoConfigTokenContentV1(parsed)) {
        throw new AutoConfigError();
    }
    return parsed;
}
exports.decodeAutoconfigTokenV1 = decodeAutoconfigTokenV1;
class AutoConfig {
    constructor(token, endpoint) {
        const content = decodeAutoconfigTokenV1(token);
        let webhook;
        try {
            webhook = new webhook_1.Webhook(content.esec);
        }
        catch (_a) {
            throw new AutoConfigError();
        }
        this.appId = content.aid;
        this.endpointId = content.eid;
        this.endpointIn = endpoint;
        this.webhook = webhook;
        const svix = new api_internal_1.SvixInternal(content.tok, { serverUrl: content.surl });
        this.requestCtx = svix.getRequestCtx();
    }
    subscribe() {
        return new endpoint_1.Endpoint(this.requestCtx).auto_config.update(this.appId, this.endpointId, {
            endpoint: this.endpointIn,
        });
    }
    verify(payload, headers) {
        return this.webhook.verify(payload, headers);
    }
}
exports.AutoConfig = AutoConfig;
