// svix/dist/autoconfigConsumer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoConfigConsumer = void 0;
const api_internal_1 = require("./api_internal");
const endpoint_1 = require("./api_internal/endpoint");
const messagePollerv2_1 = require("./api_internal/messagePollerv2");
const autoconfig_1 = require("./autoconfig");
class AutoConfigConsumer {
    constructor(token, sinkIn) {
        const content = (0, autoconfig_1.decodeAutoconfigTokenV1)(token);
        this.appId = content.aid;
        this.sinkId = content.eid;
        this.sinkIn = sinkIn;
        const svix = new api_internal_1.SvixInternal(content.tok, { serverUrl: content.surl });
        this.requestCtx = svix.getRequestCtx();
    }
    subscribe() {
        return new endpoint_1.Endpoint(this.requestCtx).auto_config.update(this.appId, this.sinkId, {
            sink: {
                type: "poller",
                config: this.sinkIn,
            },
        });
    }
    receive(consumerId, options) {
        return new messagePollerv2_1.MessagePollerv2(this.requestCtx).consumerPoll(this.appId, this.sinkId, consumerId, options);
    }
    commit(consumerId, offset, options) {
        return new messagePollerv2_1.MessagePollerv2(this.requestCtx).consumerCommit(this.appId, this.sinkId, consumerId, {
            offset,
        }, options);
    }
}
exports.AutoConfigConsumer = AutoConfigConsumer;
