// svix/dist/models/pollerV2MessageOut.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PollerV2MessageOutSerializer = void 0;
exports.PollerV2MessageOutSerializer = {
    _fromJsonObject(object) {
        return {
            channels: object["channels"],
            deliverAt: object["deliverAt"] ? new Date(object["deliverAt"]) : null,
            eventId: object["eventId"],
            eventType: object["eventType"],
            headers: object["headers"],
            id: object["id"],
            offset: object["offset"],
            payload: object["payload"],
            tags: object["tags"],
            timestamp: new Date(object["timestamp"]),
        };
    },
    _toJsonObject(self) {
        return {
            channels: self.channels,
            deliverAt: self.deliverAt,
            eventId: self.eventId,
            eventType: self.eventType,
            headers: self.headers,
            id: self.id,
            offset: self.offset,
            payload: self.payload,
            tags: self.tags,
            timestamp: self.timestamp,
        };
    },
};
