// svix/dist/models/redshiftConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedshiftConfigSerializer = void 0;
exports.RedshiftConfigSerializer = {
    _fromJsonObject(object) {
        return {
            accessKeyId: object["accessKeyId"],
            clusterIdentifier: object["clusterIdentifier"],
            dbName: object["dbName"],
            dbUser: object["dbUser"],
            region: object["region"],
            schemaName: object["schemaName"],
            secretAccessKey: object["secretAccessKey"],
            tableName: object["tableName"],
            workgroupName: object["workgroupName"],
        };
    },
    _toJsonObject(self) {
        return {
            accessKeyId: self.accessKeyId,
            clusterIdentifier: self.clusterIdentifier,
            dbName: self.dbName,
            dbUser: self.dbUser,
            region: self.region,
            schemaName: self.schemaName,
            secretAccessKey: self.secretAccessKey,
            tableName: self.tableName,
            workgroupName: self.workgroupName,
        };
    },
};
