// svix/dist/models/sinkInCommon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SinkInCommonSerializer = void 0;
exports.SinkInCommonSerializer = {
    _fromJsonObject(object) {
        return {
            channels: object["channels"],
            description: object["description"],
            disabled: object["disabled"],
            filterTypes: object["filterTypes"],
            metadata: object["metadata"],
            rateLimit: object["rateLimit"],
            secret: object["secret"],
            throttleRate: object["throttleRate"],
            uid: object["uid"],
        };
    },
    _toJsonObject(self) {
        return {
            channels: self.channels,
            description: self.description,
            disabled: self.disabled,
            filterTypes: self.filterTypes,
            metadata: self.metadata,
            rateLimit: self.rateLimit,
            secret: self.secret,
            throttleRate: self.throttleRate,
            uid: self.uid,
        };
    },
};
