// svix/dist/models/streamSinkPatch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamSinkPatchSerializer = void 0;
const amazonS3PatchConfig_1 = require("./amazonS3PatchConfig");
const azureBlobStoragePatchConfig_1 = require("./azureBlobStoragePatchConfig");
const bigQueryPatchConfig_1 = require("./bigQueryPatchConfig");
const clickhousePatchConfig_1 = require("./clickhousePatchConfig");
const eventBridgePatchConfig_1 = require("./eventBridgePatchConfig");
const googleCloudPubSubPatchConfig_1 = require("./googleCloudPubSubPatchConfig");
const googleCloudStoragePatchConfig_1 = require("./googleCloudStoragePatchConfig");
const httpPatchConfig_1 = require("./httpPatchConfig");
const otelTracingPatchConfig_1 = require("./otelTracingPatchConfig");
const rabbitMqPatchConfig_1 = require("./rabbitMqPatchConfig");
const redshiftPatchConfig_1 = require("./redshiftPatchConfig");
const sinkStatusIn_1 = require("./sinkStatusIn");
const snowflakePatchConfig_1 = require("./snowflakePatchConfig");
const snsPatchConfig_1 = require("./snsPatchConfig");
const sqsPatchConfig_1 = require("./sqsPatchConfig");
exports.StreamSinkPatchSerializer = {
    _fromJsonObject(object) {
        const type = object["type"];
        function getConfig(type) {
            switch (type) {
                case "poller":
                    return {};
                case "azureBlobStorage":
                    return azureBlobStoragePatchConfig_1.AzureBlobStoragePatchConfigSerializer._fromJsonObject(object["config"]);
                case "otelTracing":
                    return otelTracingPatchConfig_1.OtelTracingPatchConfigSerializer._fromJsonObject(object["config"]);
                case "http":
                    return httpPatchConfig_1.HttpPatchConfigSerializer._fromJsonObject(object["config"]);
                case "amazonS3":
                    return amazonS3PatchConfig_1.AmazonS3PatchConfigSerializer._fromJsonObject(object["config"]);
                case "googleCloudStorage":
                    return googleCloudStoragePatchConfig_1.GoogleCloudStoragePatchConfigSerializer._fromJsonObject(object["config"]);
                case "googleCloudPubSub":
                    return googleCloudPubSubPatchConfig_1.GoogleCloudPubSubPatchConfigSerializer._fromJsonObject(object["config"]);
                case "sqs":
                    return sqsPatchConfig_1.SqsPatchConfigSerializer._fromJsonObject(object["config"]);
                case "sns":
                    return snsPatchConfig_1.SnsPatchConfigSerializer._fromJsonObject(object["config"]);
                case "bigQuery":
                    return bigQueryPatchConfig_1.BigQueryPatchConfigSerializer._fromJsonObject(object["config"]);
                case "clickhouse":
                    return clickhousePatchConfig_1.ClickhousePatchConfigSerializer._fromJsonObject(object["config"]);
                case "eventBridge":
                    return eventBridgePatchConfig_1.EventBridgePatchConfigSerializer._fromJsonObject(object["config"]);
                case "snowflake":
                    return snowflakePatchConfig_1.SnowflakePatchConfigSerializer._fromJsonObject(object["config"]);
                case "rabbitMq":
                    return rabbitMqPatchConfig_1.RabbitMqPatchConfigSerializer._fromJsonObject(object["config"]);
                case "redshift":
                    return redshiftPatchConfig_1.RedshiftPatchConfigSerializer._fromJsonObject(object["config"]);
                default:
                    throw new Error(`Unexpected type: ${type}`);
            }
        }
        return {
            type,
            config: getConfig(type),
            batchSize: object["batchSize"],
            eventTypes: object["eventTypes"],
            maxWaitSecs: object["maxWaitSecs"],
            metadata: object["metadata"],
            status: object["status"] != null
                ? sinkStatusIn_1.SinkStatusInSerializer._fromJsonObject(object["status"])
                : undefined,
            uid: object["uid"],
        };
    },
    _toJsonObject(self) {
        let config;
        switch (self.type) {
            case "poller":
                config = {};
                break;
            case "azureBlobStorage":
                config = azureBlobStoragePatchConfig_1.AzureBlobStoragePatchConfigSerializer._toJsonObject(self.config);
                break;
            case "otelTracing":
                config = otelTracingPatchConfig_1.OtelTracingPatchConfigSerializer._toJsonObject(self.config);
                break;
            case "http":
                config = httpPatchConfig_1.HttpPatchConfigSerializer._toJsonObject(self.config);
                break;
            case "amazonS3":
                config = amazonS3PatchConfig_1.AmazonS3PatchConfigSerializer._toJsonObject(self.config);
                break;
            case "googleCloudStorage":
                config = googleCloudStoragePatchConfig_1.GoogleCloudStoragePatchConfigSerializer._toJsonObject(self.config);
                break;
            case "googleCloudPubSub":
                config = googleCloudPubSubPatchConfig_1.GoogleCloudPubSubPatchConfigSerializer._toJsonObject(self.config);
                break;
            case "sqs":
                config = sqsPatchConfig_1.SqsPatchConfigSerializer._toJsonObject(self.config);
                break;
            case "sns":
                config = snsPatchConfig_1.SnsPatchConfigSerializer._toJsonObject(self.config);
                break;
            case "bigQuery":
                config = bigQueryPatchConfig_1.BigQueryPatchConfigSerializer._toJsonObject(self.config);
                break;
            case "clickhouse":
                config = clickhousePatchConfig_1.ClickhousePatchConfigSerializer._toJsonObject(self.config);
                break;
            case "eventBridge":
                config = eventBridgePatchConfig_1.EventBridgePatchConfigSerializer._toJsonObject(self.config);
                break;
            case "snowflake":
                config = snowflakePatchConfig_1.SnowflakePatchConfigSerializer._toJsonObject(self.config);
                break;
            case "rabbitMq":
                config = rabbitMqPatchConfig_1.RabbitMqPatchConfigSerializer._toJsonObject(self.config);
                break;
            case "redshift":
                config = redshiftPatchConfig_1.RedshiftPatchConfigSerializer._toJsonObject(self.config);
                break;
        }
        return {
            type: self.type,
            config: config,
            batchSize: self.batchSize,
            eventTypes: self.eventTypes,
            maxWaitSecs: self.maxWaitSecs,
            metadata: self.metadata,
            status: self.status != null
                ? sinkStatusIn_1.SinkStatusInSerializer._toJsonObject(self.status)
                : undefined,
            uid: self.uid,
        };
    },
};
