// svix/dist/models/streamSinkIn.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamSinkInSerializer = void 0;
const azureBlobStorageConfig_1 = require("./azureBlobStorageConfig");
const bigQueryConfig_1 = require("./bigQueryConfig");
const clickhouseConfig_1 = require("./clickhouseConfig");
const eventBridgeConfig_1 = require("./eventBridgeConfig");
const googleCloudPubSubConfig_1 = require("./googleCloudPubSubConfig");
const googleCloudStorageConfig_1 = require("./googleCloudStorageConfig");
const rabbitMqConfig_1 = require("./rabbitMqConfig");
const redshiftConfig_1 = require("./redshiftConfig");
const s3Config_1 = require("./s3Config");
const sinkHttpConfig_1 = require("./sinkHttpConfig");
const sinkOtelV1Config_1 = require("./sinkOtelV1Config");
const sinkStatusIn_1 = require("./sinkStatusIn");
const snowflakeConfig_1 = require("./snowflakeConfig");
const snsConfig_1 = require("./snsConfig");
const sqsConfig_1 = require("./sqsConfig");
exports.StreamSinkInSerializer = {
    _fromJsonObject(object) {
        const type = object["type"];
        function getConfig(type) {
            switch (type) {
                case "poller":
                    return {};
                case "azureBlobStorage":
                    return azureBlobStorageConfig_1.AzureBlobStorageConfigSerializer._fromJsonObject(object["config"]);
                case "otelTracing":
                    return sinkOtelV1Config_1.SinkOtelV1ConfigSerializer._fromJsonObject(object["config"]);
                case "http":
                    return sinkHttpConfig_1.SinkHttpConfigSerializer._fromJsonObject(object["config"]);
                case "amazonS3":
                    return s3Config_1.S3ConfigSerializer._fromJsonObject(object["config"]);
                case "googleCloudStorage":
                    return googleCloudStorageConfig_1.GoogleCloudStorageConfigSerializer._fromJsonObject(object["config"]);
                case "googleCloudPubSub":
                    return googleCloudPubSubConfig_1.GoogleCloudPubSubConfigSerializer._fromJsonObject(object["config"]);
                case "sqs":
                    return sqsConfig_1.SqsConfigSerializer._fromJsonObject(object["config"]);
                case "sns":
                    return snsConfig_1.SnsConfigSerializer._fromJsonObject(object["config"]);
                case "bigQuery":
                    return bigQueryConfig_1.BigQueryConfigSerializer._fromJsonObject(object["config"]);
                case "clickhouse":
                    return clickhouseConfig_1.ClickhouseConfigSerializer._fromJsonObject(object["config"]);
                case "eventBridge":
                    return eventBridgeConfig_1.EventBridgeConfigSerializer._fromJsonObject(object["config"]);
                case "snowflake":
                    return snowflakeConfig_1.SnowflakeConfigSerializer._fromJsonObject(object["config"]);
                case "rabbitMq":
                    return rabbitMqConfig_1.RabbitMqConfigSerializer._fromJsonObject(object["config"]);
                case "redshift":
                    return redshiftConfig_1.RedshiftConfigSerializer._fromJsonObject(object["config"]);
                default:
                    throw new Error(`Unexpected type: ${type}`);
            }
        }
        return {
            type,
            config: getConfig(type),
            batchSize: object["batchSize"],
            eventTypes: object["eventTypes"],
            maxWaitSecs: object["maxWaitSecs"],
            metadata: object["metadata"],
            status: object["status"] != null
                ? sinkStatusIn_1.SinkStatusInSerializer._fromJsonObject(object["status"])
                : undefined,
            uid: object["uid"],
        };
    },
    _toJsonObject(self) {
        let config;
        switch (self.type) {
            case "poller":
                config = {};
                break;
            case "azureBlobStorage":
                config = azureBlobStorageConfig_1.AzureBlobStorageConfigSerializer._toJsonObject(self.config);
                break;
            case "otelTracing":
                config = sinkOtelV1Config_1.SinkOtelV1ConfigSerializer._toJsonObject(self.config);
                break;
            case "http":
                config = sinkHttpConfig_1.SinkHttpConfigSerializer._toJsonObject(self.config);
                break;
            case "amazonS3":
                config = s3Config_1.S3ConfigSerializer._toJsonObject(self.config);
                break;
            case "googleCloudStorage":
                config = googleCloudStorageConfig_1.GoogleCloudStorageConfigSerializer._toJsonObject(self.config);
                break;
            case "googleCloudPubSub":
                config = googleCloudPubSubConfig_1.GoogleCloudPubSubConfigSerializer._toJsonObject(self.config);
                break;
            case "sqs":
                config = sqsConfig_1.SqsConfigSerializer._toJsonObject(self.config);
                break;
            case "sns":
                config = snsConfig_1.SnsConfigSerializer._toJsonObject(self.config);
                break;
            case "bigQuery":
                config = bigQueryConfig_1.BigQueryConfigSerializer._toJsonObject(self.config);
                break;
            case "clickhouse":
                config = clickhouseConfig_1.ClickhouseConfigSerializer._toJsonObject(self.config);
                break;
            case "eventBridge":
                config = eventBridgeConfig_1.EventBridgeConfigSerializer._toJsonObject(self.config);
                break;
            case "snowflake":
                config = snowflakeConfig_1.SnowflakeConfigSerializer._toJsonObject(self.config);
                break;
            case "rabbitMq":
                config = rabbitMqConfig_1.RabbitMqConfigSerializer._toJsonObject(self.config);
                break;
            case "redshift":
                config = redshiftConfig_1.RedshiftConfigSerializer._toJsonObject(self.config);
                break;
        }
        return {
            type: self.type,
            config: config,
            batchSize: self.batchSize,
            eventTypes: self.eventTypes,
            maxWaitSecs: self.maxWaitSecs,
            metadata: self.metadata,
            status: self.status != null
                ? sinkStatusIn_1.SinkStatusInSerializer._toJsonObject(self.status)
                : undefined,
            uid: self.uid,
        };
    },
};
