// dodopayments/resources/product-collections/groups/groups.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../../core/resource.mjs";
import * as ItemsAPI from "./items.mjs";
import { Items, } from "./items.mjs";
import { buildHeaders } from "../../../internal/headers.mjs";
import { path } from "../../../internal/utils/path.mjs";
export class Groups extends APIResource {
    constructor() {
        super(...arguments);
        this.items = new ItemsAPI.Items(this._client);
    }
    /**
     * @example
     * ```ts
     * const productCollectionGroupResponse =
     *   await client.productCollections.groups.create(
     *     'pdc_8BWv0hojwUH7iCDabr0NI',
     *     { products: [{ product_id: 'product_id' }] },
     *   );
     * ```
     */
    create(id, body, options) {
        return this._client.post(path `/product-collections/${id}/groups`, { body, ...options });
    }
    /**
     * @example
     * ```ts
     * await client.productCollections.groups.delete(
     *   '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   { id: 'pdc_8BWv0hojwUH7iCDabr0NI' },
     * );
     * ```
     */
    delete(groupID, params, options) {
        const { id } = params;
        return this._client.delete(path `/product-collections/${id}/groups/${groupID}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * @example
     * ```ts
     * await client.productCollections.groups.update(
     *   '182bd5e5-6e1a-4fe4-a799-aa6d9a6ab26e',
     *   { id: 'pdc_8BWv0hojwUH7iCDabr0NI' },
     * );
     * ```
     */
    update(groupID, params, options) {
        const { id, ...body } = params;
        return this._client.patch(path `/product-collections/${id}/groups/${groupID}`, {
            body,
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
}
Groups.Items = Items;
